---
title: "Grid Modularity"
output: 
  html_document:
    toc: true
    toc_float: true
    collapse: false
    number_sections: true
    code_folding: hide
    theme: united
    highlight: tango
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, results = "asis", message=FALSE, warning=FALSE, cache=TRUE, eval = TRUE, dev = c('png'), out.width = '100%', out.height='40%')
```


```{r load libraries, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(magrittr)
library(ggplot2)
library(vegan)
library(igraph)
library(reshape2)
library(infomapecology)
library(aricode)

rm(list=ls())
```

```{r}
# reading ASV data
data_asv <- read_csv("data_processed/three_villages/data_asv_0.01_grids2.csv") 
data_asv %<>% filter((host_species == "Rattus rattus" | host_species == "Microgale brevicaudata") & village == "Mandena") %>% 
  mutate(grid = factor(grid, levels = c("semi-intact_forest","secondary_forest","brushy_regrowth","agriculture","flooded_rice","agroforest","village")))

# ASVs taxonomy
asv_taxa <- read_delim("data_raw/ASVs_all_merged_taxonomy.tsv") %>% 
  select(ASV, Family)
data_asv %<>% left_join(asv_taxa, by=c("asv_ID"="ASV")) 

data_asv_family <- data_asv %>% 
  group_by(host_ID, village, host_species, grid, season, mass, sex, dist_village, Family) %>% 
  summarise(reads = sum(reads)) %>% 
  filter(!(is.na(Family)))

n_asv <- data_asv %>%
  count(asv_ID) %>%
  filter(n<3)
#
#data_asv %<>% filter(asv_ID %in% n_asv$asv_ID)

#data_asv_mat <- data_asv %>% 
#  spread(asv_ID, reads, fill = 0) %>% 
#  mutate(across(starts_with("ASV"), ~.*unfiltered_reads)) %>% 
#  mutate(total_reads = select(.,starts_with("ASV")) %>% rowSums(na.rm = TRUE)) %>% 
#  mutate(across(starts_with("ASV"), ~./total_reads)) %>% 
#  gather("asv_ID", "reads", starts_with("ASV")) %>% 
#  filter(reads>0)

#data_mammals <- read_csv("data_raw/Terrestrial_Mammals.csv")
#data_mammals %<>% mutate(host_ID = as.numeric(gsub(".*?([0-9]+).*", "\\1", animal_id))) %>% 
#  mutate(dist_village = as.numeric(gsub(".*?([0-9,.]+).*", "\\1", dist_village_center))) %>%
#  select(host_ID, dist_village) 

#full_grids <- read_csv("data_processed/three_villages/data_SM.csv") %>% dplyr::select(host_ID, grid) %>% 
#  left_join(data_mammals, by="host_ID")
#data_asv %<>% dplyr::select(-grid) %>%left_join(full_grids, by="host_ID") %>% 
#  mutate(dist_village = ifelse(grid=="village",0,dist_village))
#write_csv(data_asv_mat, "data_processed/three_villages/data_asv_0.01_grids2.csv")

# distance between grids
grid_dist_avg <- data_asv %>% 
  distinct(host_ID, grid, dist_village) %>% 
  group_by(grid) %>% 
  summarise(distance = mean(dist_village))

data_asv_mat <- data_asv %>% 
  mutate(reads = 1) %>% 
  spread(asv_ID, reads, fill = 0) %>% 
  column_to_rownames("host_ID") %>% 
  dplyr::select(starts_with("ASV")) %>% 
  as.matrix()
```

# Interactions level

## Microbes diversity {.tabset}
\
\

### Alpha-diversity
\
\
```{r}
richness <- data_asv %>% group_by(host_ID) %>% 
  summarise(richness = n_distinct(asv_ID))

#plotting ASVs richness by village, grid, and host species
g <- data_asv %>% 
  group_by(village, grid, host_species, host_ID) %>% 
  summarise(richness = n_distinct(asv_ID)) %>% 
  ggplot(aes(x=grid, y=richness, fill=grid)) +
  geom_boxplot() +
  facet_wrap(~host_species)+
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black', angle = 45, vjust = 0.5, hjust=1), title = element_text(size = 16), strip.text.x = element_text(size = 12), legend.position = "none") +
  labs(x="Land-use", y="Average ASV richness") 
print(g)
cat('\n','\n')
```

### Beta-diversity
\
\
```{r}
# similarity Jaccard

host_attr <- data_asv %>% 
  distinct(host_ID, host_species, grid)

# calculating distance
jaccard_dist <- as.matrix(1-vegdist(data_asv_mat, method = "jaccard"))
# removing duplications
jaccard_dist[lower.tri(jaccard_dist)] <- NA
diag(jaccard_dist) <- NA
# long format + adding host attributes
jaccard_m <- melt(jaccard_dist) %>% 
  filter(!(is.na(value))) %>%  
  left_join(host_attr, by=c("Var1"="host_ID")) %>% 
  dplyr::rename(host_species1 = host_species, grid1=grid) %>% 
  left_join(host_attr, by=c("Var2"="host_ID")) %>% 
  dplyr::rename(host_species2 = host_species, grid2=grid) %>% 
  mutate(same_species = host_species1==host_species2)

# t.test between species
jac_same_species <- jaccard_m %>% filter(same_species==TRUE)
jac_results <- t.test(value~host_species1, data=jac_same_species)
cat("t = ", jac_results$statistic, "p = ", jac_results$p.value)
cat('\n','\n')

# plotting
g <- jac_same_species %>% 
  ggplot(aes(x=host_species1, y=value, fill=host_species1)) + 
  geom_boxplot() + 
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14)) +
  labs(x="Host Species", y="Microbiome Similarity [Jaccard]")
print(g)
cat('\n','\n')
```

### PERMANOVA
\
\
```{r}
cat("Rattus rattus")
cat('\n','\n')
data_asv_rattus <- data_asv %>% filter(host_species=="Rattus rattus")
rattus_attr <- data_asv_rattus %>% distinct(host_ID, host_species, grid, season, dist_village)
# matrix
data_asv_rattus_mat <- data_asv_rattus %>% 
  mutate(reads = 1) %>% 
  spread(asv_ID, reads, fill = 0) %>% 
  column_to_rownames("host_ID") %>% 
  dplyr::select(starts_with("ASV")) %>% 
  as.matrix()
# PERMANOVA
perm_ratus <- as.data.frame(adonis2(formula = data_asv_rattus_mat ~ grid, data = rattus_attr,  method = "jaccard"))
print(knitr::kable(perm_ratus[1,]))

# NMDS
nmds_rattus <- metaMDS(data_asv_rattus_mat, distance = "jaccard", noshare = TRUE, k=2, trace = FALSE)
NMDS1<-nmds_rattus$points[,1]
NMDS2<-nmds_rattus$points[,2]
rattus_plot <- cbind(rattus_attr, NMDS1, NMDS2)

# plotting
g <- rattus_plot %>% 
  ggplot( aes(NMDS1, NMDS2, color=grid)) +
  geom_point(position=position_jitter(.1)) +
  stat_ellipse(aes(fill=grid), alpha=.1, type='norm',linetype =2, geom="polygon") + ##draws 95% confidence interval ellipses
  theme_minimal() +
  annotate("text", x=0, y=max(abs(NMDS2)), label=paste('Stress =',round(nmds_rattus$stress,3))) # adding stress
print(g)
cat('\n','\n')


cat("Microgale brevicaudata")
cat('\n','\n')
data_asv_microgale <- data_asv %>% filter(host_species=="Microgale brevicaudata")
microgale_attr <- data_asv_microgale %>% distinct(host_ID, host_species, grid, season, dist_village)
# matrix
data_asv_microgale_mat <- data_asv_microgale %>% 
  mutate(reads = 1) %>% 
  spread(asv_ID, reads, fill = 0) %>% 
  column_to_rownames("host_ID") %>% 
  dplyr::select(starts_with("ASV")) %>% 
  as.matrix()
# PERMANOVA
perm_microgale <- as.data.frame(adonis2(formula = data_asv_microgale_mat ~ grid, data = microgale_attr,  method = "jaccard"))
print(knitr::kable(perm_microgale[1,]))

# NMDS
nmds_microgale <- metaMDS(data_asv_microgale_mat, distance = "jaccard", noshare = TRUE, k=2, trace = FALSE)
NMDS1<-nmds_microgale$points[,1]
NMDS2<-nmds_microgale$points[,2]
microgale_plot <- cbind(microgale_attr, NMDS1, NMDS2)

# plotting
g <- microgale_plot %>% 
  ggplot( aes(NMDS1, NMDS2, color=grid)) +
  geom_point(position=position_jitter(.1)) +
  stat_ellipse(aes(fill=grid), alpha=.1, type='norm',linetype =2, geom="polygon") + ##draws 95% confidence interval ellipses
  theme_minimal() +
  annotate("text", x=0, y=max(abs(NMDS2)), label=paste('Stress =',round(nmds_microgale$stress,3))) # adding stress
print(g)
cat('\n','\n')
```

### Degree distribution
\
\
```{r}
# connectance
# possible and realized edges
data_asv_rattus <- data_asv %>% filter(host_species=="Rattus rattus") 
n_pos_edges_rattus <- length(unique(data_asv_rattus$host_ID)) * length(unique(data_asv_rattus$asv_ID))
n_edges_rattus <- nrow(data_asv_rattus)
data_asv_microgale <- data_asv %>% filter(host_species=="Microgale brevicaudata") 
n_pos_edges_microgale <- length(unique(data_asv_microgale$host_ID)) * length(unique(data_asv_microgale$asv_ID))
n_edges_microgale <- nrow(data_asv_microgale)

conect <- tibble(host_species = unique(data_asv$host_species),
       connectance = c(n_edges_rattus/n_pos_edges_rattus, n_edges_microgale/n_pos_edges_microgale))
  print(knitr::kable(conect))
cat('\n','\n')

g <- data_asv %>% 
  group_by(host_species, asv_ID) %>% 
  summarise(prevalence = n_distinct(host_ID)) %>%
  ggplot(aes(x=prevalence)) +
  geom_histogram(binwidth = 1) +
  facet_wrap(~host_species)+
  geom_vline(xintercept = c(10), linetype="dashed") +
  xlim(0,100)+
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14), strip.text.x = element_text(size=12)) +
  labs(x="ASVs Degree", y="No. of ASVs")
print(g)
cat('\n','\n')
```

### Taxonomy
\
\
```{r}
# ASVs taxonomy
asv_taxa <- read_delim("data_raw/ASVs_all_merged_taxonomy.tsv")

# average ASVs relative abundance across grids
asv_rel_abu_grid <- data_asv %>% 
  group_by(asv_ID, grid, host_species) %>% 
  summarise(rel_abundance = mean(reads)) %>% 
  left_join(asv_taxa, by=c("asv_ID"="ASV"))

n_phylum <- length(unique(asv_rel_abu_grid$Phylum)) 
# relative abundance by Phylum
g <- asv_rel_abu_grid %>% 
  group_by(grid, host_species, Phylum) %>% 
  summarise(rel_abundance = mean(rel_abundance)) %>% 
  ggplot(aes(x=grid, y=rel_abundance, fill=Phylum)) + 
  geom_bar(position="fill", stat="identity") + 
  facet_wrap(~host_species) +
  theme_bw()+
  scale_fill_manual(values = as.character(pals::alphabet2(n=n_phylum)))+
  theme(axis.text = element_text(size = 10, color = 'black', angle = 45, vjust = 0.5, hjust=1), title = element_text(size = 14), strip.text.x = element_text(size = 12)) +
  labs(x="Land Use", y = "ASVs Average Relative Abundance")
print(g)
cat('\n','\n')
```


## Mantel-test {.tabset}
\
\

### Rattus rattus
\
\
```{r}
# Jaccard
rattus_jaccard <- 1-vegdist(data_asv_rattus_mat, method = "jaccard")

# distance between grids
  rattus_dist <- expand_grid(id1=rattus_attr$host_ID,id2=rattus_attr$host_ID) %>% 
    left_join(rattus_attr, by=c("id1"="host_ID")) %>% rename(dist1=dist_village) %>% 
    left_join(rattus_attr, by=c("id2"="host_ID")) %>% rename(dist2=dist_village) %>% 
    mutate(distance = abs(dist1-dist2)) %>% 
    select(id1,id2,distance) %>% 
    spread(id2, distance) %>% 
    column_to_rownames("id1") %>% 
    as.dist()
  
  # mantel-test
  mantel_rattus <- ecodist::mantel(rattus_jaccard ~ rattus_dist)
  # printing the results
mantel_rattus_results <- tibble(r = mantel_rattus[1],
                              pvalue = mantel_rattus[3]) 
  print(knitr::kable(mantel_rattus_results))
cat('\n','\n')

# plotting the correlation
rattus_jaccard_m <- as.matrix(rattus_jaccard)
rattus_jaccard_m[lower.tri(rattus_jaccard_m)] <- NA
diag(rattus_jaccard_m) <- NA
rattus_jaccard_m <- melt(rattus_jaccard_m) %>% 
  filter(!is.na(value))

rattus_dist_m <- melt(as.matrix(rattus_dist)) %>% rename(distance=value)

g <- rattus_jaccard_m %>% 
  left_join(rattus_dist_m, by=c("Var1","Var2")) %>% 
ggplot(aes(y=value, x=distance)) +
  geom_point(alpha = 0.5, position=position_jitter(width=.1,height=.01)) +
  geom_smooth(method = "glm", se=T, method.args = list(family = "gaussian")) +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14), strip.text.x = element_text(size=12)) +
  labs(x = "Distance Between Individuals [km]", y = "ASVs Similarity [Jaccard]")
print(g)
  cat('\n','\n')
```


### Microgale brevicaudata
\
\
```{r}
# Jaccard
microgale_jaccard <- 1-vegdist(data_asv_microgale_mat, method = "jaccard")

# distance between grids
  microgale_dist <- expand_grid(id1=microgale_attr$host_ID,id2=microgale_attr$host_ID) %>% 
    left_join(microgale_attr, by=c("id1"="host_ID")) %>% rename(dist1=dist_village) %>% 
    left_join(microgale_attr, by=c("id2"="host_ID")) %>% rename(dist2=dist_village) %>% 
    mutate(distance = abs(dist1-dist2)) %>% 
    select(id1,id2,distance) %>% 
    spread(id2, distance) %>% 
    column_to_rownames("id1") %>% 
    as.dist()
  
  # mantel-test
  mantel_microgale <- ecodist::mantel(microgale_jaccard ~ microgale_dist)
  # printing the results
mantel_microgale_results <- tibble(r = mantel_microgale[1],
                              pvalue = mantel_microgale[3]) 
  print(knitr::kable(mantel_microgale_results))
cat('\n','\n')

# plotting the correlation
microgale_jaccard_m <- as.matrix(microgale_jaccard)
microgale_jaccard_m[lower.tri(microgale_jaccard_m)] <- NA
diag(microgale_jaccard_m) <- NA
microgale_jaccard_m <- melt(microgale_jaccard_m) %>% 
  filter(!is.na(value))

microgale_dist_m <- melt(as.matrix(microgale_dist)) %>% rename(distance=value)

g <- microgale_jaccard_m %>% 
  left_join(microgale_dist_m, by=c("Var1","Var2")) %>% 
ggplot(aes(y=value, x=distance)) +
  geom_point(alpha = 0.5, position=position_jitter(width=.1,height=.01)) +
  geom_smooth(method = "glm", se=T, method.args = list(family = "gaussian")) +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14), strip.text.x = element_text(size=12)) +
  labs(x = "Distance Between Individuals [km]", y = "ASVs Similarity [Jaccard]")
print(g)
  cat('\n','\n')
```



# Structural level - Host {.tabset}
\
\

```{r}
for(s in c("Rattus rattus", "Microgale brevicaudata")) {
  cat('##',s,'{.tabset}','\n','\n')
  
  data_asv_s <- data_asv %>% filter(host_species==s) 
    #filter(host_ID %in% b$host_ID)
  
  data_asv_mat <- data_asv_s %>% 
  mutate(reads = 1) %>% 
  spread(asv_ID, reads, fill = 0) %>% 
  column_to_rownames("host_ID") %>% 
  dplyr::select(starts_with("ASV")) %>% 
  as.matrix()
  
  data_asv_mat_family <- data_asv_s %>% 
  #mutate(reads = 1) %>% 
  spread(Family, reads, fill = 0) %>% 
  column_to_rownames("host_ID") %>% 
  dplyr::select(-village, -host_species, -grid, -season, -mass, -sex, -dist_village) %>% 
  as.matrix()

  # building the network
network_object <- infomapecology::create_monolayer_object(data_asv_mat, directed = FALSE, bipartite = TRUE, group_names = c("ASV", "Host"))

# modularity analysis
infomap_object <- run_infomap_monolayer(network_object,
                                        infomap_executable='Infomap',
                                        flow_model = 'undirected',
                                        two_level = TRUE,
                                        silent=TRUE, trials=100, seed=123)
cat('\n','\n')

# adding the modules classification to the data
modules_asv <- infomap_object$modules %>% filter(node_group == "ASV")%>% dplyr::rename(asv_group = module_level1, asv_ID = node_name)
modules_host <- infomap_object$modules %>% filter(node_group == "Host")%>% dplyr::rename(host_group = module_level1, host_ID = node_name) %>% mutate(host_ID = as.numeric(host_ID))

data_asv_s %<>% left_join(modules_asv %>% dplyr::select(asv_ID,asv_group), by = "asv_ID") %>% 
  left_join(modules_host %>% dplyr::select(host_ID,host_group), by = "host_ID")

# shuffled data
network_data1 <- read_csv("data_processed/shuffled/network_data_rattus_curveball_100")
network_data2 <- read_csv("data_processed/shuffled/network_data_microgale_curveball_100")
network_data_shuff <- rbind(network_data1, network_data2)

modules_data1 <- read_csv("data_processed/shuffled/modules_sim_data_rattus_curveball_100")
modules_data2 <- read_csv("data_processed/shuffled/modules_sim_data_microgale_curveball_100")
modules_data_shuff <- rbind(modules_data1, modules_data2)

###
cat('###', "Modules exploration",'\n')
# number of modules
cat("Number of modules: ", length(unique(data_asv_s$host_group)))
cat('\n','\n')

#  modules
n_host_grid <- data_asv_s %>% 
  distinct(host_ID,grid) %>% 
  count(grid) %>% 
  dplyr::rename(n_grid=n)
  
g <- data_asv_s %>% 
  distinct(host_ID, grid, host_group) %>% 
  count(grid, host_group) %>% 
  left_join(n_host_grid, by="grid") %>% 
  mutate(n_rel = n/n_grid) %>% 
  ggplot(aes(x = host_group, y = grid, fill=n_rel)) +
  geom_tile(color='white') +
  theme_classic() +
  scale_fill_viridis_c(limits = c(0, 1)) +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14), strip.text.x = element_text(size=12)) +
  labs(x='Module ID', y='Land Use')
print(g)
cat('\n','\n')

# modules size
g <- data_asv_s %>% 
  group_by(host_group) %>% 
  summarise(n = n_distinct(host_ID)) %>% 
  ggplot(aes(x=n)) +
  geom_histogram(binwidth = 1) +
  theme_bw() +
  theme(axis.text = element_text(size = 14, color = 'black'), title = element_text(size = 20), strip.text.x = element_text(size=12)) +
  labs(x="Module Size", y="No. of Modules")
print(g)
cat('\n','\n')

# in how many land uses modules occur?
n_landuse <- data_asv_s %>% 
  group_by(host_group) %>% 
  summarise(n = n_distinct(grid))  
 g <- n_landuse %>%  ggplot(aes(x=n)) +
  geom_histogram(binwidth = 1) +
  theme_bw() +
  theme(axis.text = element_text(size = 14, color = 'black'), title = element_text(size = 20), strip.text.x = element_text(size=12)) +
  labs(x="No. of Land Uses", y="No. of Modules")
print(g)
cat('\n','\n')

# 

###
cat('###', "No. of modules",'\n')

# observed number of modules
obs_m <- infomap_object$m

# shuffled number of modules
shuff_m <- network_data_shuff %>% filter(host_species==s)

# plotting
p <- nrow(shuff_m %>% filter(m<=obs_m)) / nrow(shuff_m)
g <- shuff_m %>% 
  ggplot(aes(x=m)) + 
  geom_histogram() + 
  theme_bw() +
  geom_vline(xintercept = obs_m, linetype='dashed', color="red") +
  annotate("text", x=obs_m+3, y=15, label=paste('p-value =',p)) +
  labs(x = "No. of Modules", y = "No. of Shuffled Networks")
print(g)
cat('\n','\n')

###
cat('###', "Mantel-test for modules similarity",'\n')

## observed network
# matrix of grid similarity in modules
grid_modules <- data_asv_s %>% 
  group_by(grid, host_group) %>% 
  summarise(host_n = n_distinct(host_ID)) %>% 
  spread(host_group, host_n, fill = 0) %>% 
  mutate(grid = as.character(grid)) %>%
  arrange(grid) %>% 
  column_to_rownames("grid") %>% 
  as.matrix() 

grid_modules_dist <- as.matrix(1-vegdist(sqrt(grid_modules), method = "bray"))

# transforming to long format
grid_mudules_dist_m <- grid_modules_dist
grid_mudules_dist_m[lower.tri(grid_mudules_dist_m)] <- NA
diag(grid_mudules_dist_m) <- NA
grid_mudules_dist_m <- melt(grid_mudules_dist_m) %>% 
  filter(!is.na(value)) %>% 
  dplyr::rename(mean=value) %>% 
  mutate(sd = NA, type="Observed")

# matrix of distances between grids
grid_attr <- data_asv_s %>% 
  distinct(grid) %>%
 left_join(grid_dist_avg, by="grid")

grid_dist <- expand_grid(id1=grid_attr$grid,id2=grid_attr$grid) %>% 
    left_join(grid_attr, by=c("id1"="grid")) %>% dplyr::rename(dist1=distance) %>% 
    left_join(grid_attr, by=c("id2"="grid")) %>% dplyr::rename(dist2=distance) %>% 
    mutate(distance = abs(dist1-dist2)) %>% 
    select(id1,id2,distance) %>% 
    spread(id2, distance) %>% 
    column_to_rownames("id1") %>% 
    as.matrix()
# long format
grid_dist_m <- melt(grid_dist) %>% dplyr::rename(distance=value)

## shuffled networks
modules_shuff_avg <- modules_data_shuff %>% 
  filter(host_species==s) %>% 
  group_by(Var1, Var2) %>% 
  summarise(mean = mean(value), sd = sd(value)) %>% 
  mutate(type="Shuffled")

# mantel test observed
mantel_grid_modules <- ecodist::mantel(as.dist(grid_modules_dist) ~ as.dist(grid_dist))

# mantel test shuffled
grid_modules_dist_shuff1 <- modules_shuff_avg %>% 
  dplyr::select(Var1, Var2, mean) 
grid_modules_dist_shuff2 <- grid_modules_dist_shuff1 %>% dplyr::rename(Var1=Var2, Var2=Var1)
grid_modules_dist_shuff <- rbind(grid_modules_dist_shuff1,grid_modules_dist_shuff2) %>% 
  spread(Var2, mean, fill = 0) %>% 
  column_to_rownames("Var1") %>% 
  as.matrix()
mantel_grid_modules_shuff <- ecodist::mantel(as.dist(grid_modules_dist_shuff) ~ as.dist(grid_dist))

# printing the results
mantel_grid_results <- tibble(type = c("Observed","shuffled"),
                              r = c(mantel_grid_modules[1], mantel_grid_modules_shuff[1]),
                              pvalue = c(mantel_grid_modules[3], mantel_grid_modules_shuff[3])) 
  print(knitr::kable(mantel_grid_results))
cat('\n','\n')

# plotting the correlation
final_grid_modules <- rbind(grid_mudules_dist_m, modules_shuff_avg)
g <- final_grid_modules %>% 
  left_join(grid_dist_m, by=c("Var1","Var2")) %>% 
ggplot(aes(y=mean, x=distance, color=type)) +
  geom_point(alpha = 0.8) +
  geom_errorbar(aes(ymin= mean-sd, ymax= mean+sd))+
  geom_smooth(method = "glm", se=F, method.args = list(family = "gaussian")) +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14), strip.text.x = element_text(size=12)) +
  labs(x = "Distance Between Grids [km]", y = "Modules Similarity [Bray-Curtis]")
print(g)
  cat('\n','\n')
  
  # plotting the correlation
final_grid_modules <- grid_mudules_dist_m %>% left_join(modules_shuff_avg, by=c("Var1","Var2")) %>% mutate(z = (mean.x-mean.y)/sd.y) %>% 
  mutate(sig = ifelse(z>1.96,"red", ifelse(z<(-1.96),"blue","black")))
g <- final_grid_modules %>% 
  left_join(grid_dist_m, by=c("Var1","Var2")) %>% 
ggplot(aes(y=mean.x, x=distance)) +
  geom_point(color=final_grid_modules$sig) +
  geom_smooth(method = "glm", se=F, method.args = list(family = "gaussian")) +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14), strip.text.x = element_text(size=12)) +
  labs(x = "Distance Between Grids [km]", y = "Modules Similarity [Bray-Curtis]")
print(g)
  cat('\n','\n')

  ###
cat('###', "Dendrogram",'\n')

# modules similarity between grids as tree
# calculate distance by UPGMA
phy <- as.dist(1-grid_modules_dist) %>%
  hclust(method = "average") 
print(plot(phy, main = NULL, xlab = NULL, ylab = "Modules Dissimilarity", ann=FALSE))
  cat('\n','\n')

  ###
cat('###', "NMI",'\n')
  hosts <- data_asv_s %>% distinct(host_ID, grid, host_group) 

nmi_obs <- aricode::NMI(hosts$grid, hosts$host_group)
nmi_shuff <- vector(length = 1000)
for(i in 1:1000) {
  # shuffling the grid attribute
  hosts_shuff <- hosts %>% 
    mutate(grid = sample(grid,nrow(hosts)))
  
  # calculating nmi
  nmi_shuff[i] <- aricode::NMI(hosts_shuff$grid, hosts_shuff$host_group)
}

# plotting
p <- length(nmi_shuff[nmi_shuff>nmi_obs]) / length(nmi_shuff)
g <- as.data.frame(nmi_shuff) %>% 
  ggplot(aes(nmi_shuff)) + 
  geom_histogram() + 
  theme_bw() +
  geom_vline(xintercept = nmi_obs, linetype='dashed', color="red") +
  theme(axis.text = element_text(size = 14, color = 'black'), title = element_text(size = 20)) +
  labs(x="NMI", y="No. of Shuffled Networks")+
  annotate("text", x=max(nmi_obs), y=100, label=paste('p-value =',p))
print(g)
cat('\n','\n')

}
```

```{r}
n_module_grid <- data_asv_s %>% 
  group_by(host_group) %>% 
  summarise(n_grid = n_distinct(grid))

#a=quantile(n_asv$n, c(0.25,0.75))

n_asv <- data_asv_s %>% 
  count(asv_ID) %>% 
  left_join(modules_asv %>% dplyr::select(asv_ID, asv_group), by="asv_ID") %>% 
  left_join(n_module_grid, by = c("asv_group"="host_group")) %>% 
  mutate(n = ifelse(n<5,"1-4", ifelse(n>=10,"10+", "5-9"))) %>% 
  mutate(n = factor(n, levels = c("1-4","5-9","10+"))) 

n_asv %>% 
  ggplot(aes(x=n, y=n_grid, fill=n)) + 
  geom_boxplot() + 
  theme_bw() +
  theme(axis.text = element_text(size = 14, color = 'black'), title = element_text(size = 20), legend.position = "none") +
  labs(x="ASVs Degree", y="Modules' No. of Land Uses")
ggsave("n_grids_modules.pdf", path = "grid_modularity_files", device = "pdf")

```

```{r}
n_landuse <- data_asv_s %>% 
  group_by(host_group) %>% 
  summarise(n = n_distinct(grid))

n_asv <- data_asv_s %>% 
  count(asv_ID)

n_module <- data_asv_s %>% 
  group_by(host_group) %>% 
  summarise(module_size = n_distinct(host_ID))

n_host_asv <- data_asv_s %>%
  group_by(host_ID) %>% 
  summarise(richness = n_distinct(asv_ID))

dat <- data_asv_s %>% 
  left_join(n_landuse, by="host_group") %>% 
  rename( n= "n_grid" ) %>% 
  mutate(n_grid = as.factor(n_grid)) %>%  
  left_join(n_asv, by="asv_ID") %>% 
  rename(n="n_asv" ) %>% 
  left_join(n_host_asv, by="host_ID") %>% 
  left_join(n_module, by="host_group")

host_n_grid <- dat %>% 
  group_by(n_grid, host_ID, mass, sex, module_size) %>% 
  summarise(n = mean(n_asv), richness=mean(richness))

host_n_grid_summary <- host_n_grid %>% 
  group_by(n_grid) %>% 
  summarise(n = mean(n))

# microbe average prevalence
host_n_grid %>% 
  ggplot(aes(x=n_grid, y=n, fill=n_grid)) + 
  geom_boxplot() + 
  theme_bw() +
  theme(axis.text = element_text(size = 14, color = 'black'), title = element_text(size = 20), legend.position = "none") +
  labs(x="No. of Land Uses", y="Average Microbe Prevalence")

# host average richness
host_n_grid %>% 
  ggplot(aes(x=n_grid, y=richness, fill=n_grid)) + 
  geom_boxplot() + 
  theme_bw() +
  theme(axis.text = element_text(size = 14, color = 'black'), title = element_text(size = 20), legend.position = "none") +
  labs(x="No. of Land Uses", y="Average Host Richness")

# host average mass
host_n_grid_sex <- host_n_grid %>% 
  #dplyr::filter(sex == "male") %>% 
  group_by(n_grid) %>% 
  summarise(n = mean(mass))
 host_n_grid %>%  ggplot(aes(x=n_grid, y=mass, fill=n_grid)) + 
  geom_boxplot() + 
  theme_bw() +
  theme(axis.text = element_text(size = 14, color = 'black'), title = element_text(size = 20), legend.position = "none") +
  labs(x="No. of Land Uses", y="Average Microbe Prevalence")

 # host age
 data_mammals <- read_csv("data_raw/Terrestrial_Mammals.csv")
 data_mammals %<>% mutate(host_ID = as.numeric(gsub(".*?([0-9]+).*", "\\1", animal_id))) %>% 
   select(host_ID, age_repro) 
  #data_mammals[data_mammals==3] <- 2
 
 host_n_grid <- host_n_grid %>% 
   left_join(data_mammals, by="host_ID") 
 
 host_n_grid_age <- host_n_grid %>% 
   filter(n>75 | n < 15) %>% 
   #filter(sex == "male") %>% 
  group_by(n_grid) %>% 
  summarise(n = mean(age_dental))
 
 host_n_grid %>%  ggplot(aes(x=module_size)) + 
  geom_histogram() + 
  theme_bw() 
 a=quantile(host_n_grid$module_size, c(0.25,0.75))
 
 host_n_grid_age2 <- host_n_grid %>% 
   mutate(n_grid = as.double(n_grid)) %>%
   mutate(pre = ifelse(n_grid<2,"low", ifelse(n_grid>5,"high", "medium"))) %>% 
  mutate(pre = factor(pre, levels = c("low","medium","high"))) %>% 
   #filter(sex == "male") %>% 
  group_by(pre) %>% 
   count(sex)
  summarise(n = mean(richness))
 
 host_n_grid %>% 
   mutate(pre = ifelse(module_size<3,"low", ifelse(module_size>100,"high", "medium"))) %>% 
  mutate(pre = factor(pre, levels = c("low","medium","high"))) %>% 
   mutate(n_grid = as.double(n_grid)) %>% 
   #filter(sex == "male") %>%
  ggplot(aes(x=pre, y=age_repro, fill=pre)) + 
  geom_violin() + 
  theme_bw() +
  theme(axis.text = element_text(size = 14, color = 'black'), title = element_text(size = 20), legend.position = "none") +
  labs(x="ASVs Degree", y="Modules' No. of Land Uses")
 
 #b=host_n_grid_age2 %>% select(host_ID,pre) %>% filter(pre!="high")
 
 # NMI
 hosts <- data_asv_s %>% 
   left_join(select(host_n_grid_age2, host_ID,pre), by="host_ID") %>% filter(pre!="low") %>% 
   distinct(host_ID, grid, host_group) 
nmi_obs <- aricode::NMI(hosts$grid, hosts$host_group)
nmi_shuff <- vector(length = 1000)
for(i in 1:1000) {
  # shuffling the grid attribute
  hosts_shuff <- hosts %>% 
    mutate(grid = sample(grid,nrow(hosts)))
  
  # calculating nmi
  nmi_shuff[i] <- aricode::NMI(hosts_shuff$grid, hosts_shuff$host_group)
}

# plotting
p <- length(nmi_shuff[nmi_shuff>nmi_obs]) / length(nmi_shuff)
```

```{r}
host_g <- dat %>% 
  group_by(n_grid, host_ID, module_size) %>% 
  summarise(n = mean(n_asv), richness=mean(richness)) %>% 
  mutate(n_grid = as.double(n_grid)) %>% 
  column_to_rownames("host_ID") %>% 
  as.matrix()

# NMDS
nmds_rattus <- metaMDS(host_g, distance = "bray", noshare = TRUE, k=2, trace = FALSE)
NMDS1<-nmds_rattus$points[,1]
NMDS2<-nmds_rattus$points[,2]
rattus_plot <- cbind(host_n_grid, NMDS1, NMDS2)

# plotting
g <- rattus_plot %>% 
  ggplot( aes(NMDS1, NMDS2, color=as.factor(age_repro))) +
  geom_point(position=position_jitter(.1)) +
  #stat_ellipse(aes(fill=grid), alpha=.1, type='norm',linetype =2, geom="polygon") + ##draws 95% confidence interval ellipses
  theme_minimal() 
  #annotate("text", x=0, y=max(abs(NMDS2)), label=paste('Stress =',round(nmds_rattus$stress,3))) # adding stress
print(g)

```


# Structural level - ASVs {.tabset}
\
\

```{r}
for(s in c("Rattus rattus", "Microgale brevicaudata")) {
  cat('##',s,'{.tabset}','\n','\n')
  
  data_asv_s <- data_asv %>% filter(host_species==s)
  
  data_asv_mat <- data_asv_s %>% 
  mutate(reads = 1) %>% 
  spread(asv_ID, reads, fill = 0) %>% 
  column_to_rownames("host_ID") %>% 
  dplyr::select(starts_with("ASV")) %>% 
  as.matrix()

  # building the network
network_object <- infomapecology::create_monolayer_object(data_asv_mat, directed = FALSE, bipartite = TRUE, group_names = c("ASV", "Host"))

# modularity analysis
infomap_object <- run_infomap_monolayer(network_object,
                                        infomap_executable='Infomap',
                                        flow_model = 'undirected',
                                        two_level = TRUE,
                                        silent=TRUE, trials=100, seed=123)
cat('\n','\n')

# adding the modules classification to the data
modules_asv <- infomap_object$modules %>% filter(node_group == "ASV")%>% dplyr::rename(asv_group = module_level1, asv_ID = node_name)
modules_host <- infomap_object$modules %>% filter(node_group == "Host")%>% dplyr::rename(host_group = module_level1, host_ID = node_name) %>% mutate(host_ID = as.numeric(host_ID))

data_asv_s %<>% left_join(modules_asv %>% dplyr::select(asv_ID,asv_group), by = "asv_ID") %>% 
  left_join(modules_host %>% dplyr::select(host_ID,host_group), by = "host_ID")


###
cat('###', "Modules exploration",'\n')
# number of modules
cat("Number of modules: ", length(unique(data_asv_s$asv_group)))
cat('\n','\n')

#  modules
n_host_grid <- data_asv_s %>% 
  distinct(asv_ID,grid) %>% 
  count(grid) %>% 
  dplyr::rename(n_grid=n)
  
g <- data_asv_s %>% 
  distinct(asv_ID, grid, asv_group) %>% 
  count(grid, asv_group) %>% 
  left_join(n_host_grid, by="grid") %>% 
  mutate(n_rel = n/n_grid) %>% 
  ggplot(aes(x = asv_group, y = grid, fill=n_rel)) +
  geom_tile(color='white') +
  theme_classic() +
  scale_fill_viridis_c(limits = c(0, 1)) +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14), strip.text.x = element_text(size=12)) +
  labs(x='Module ID', y='Land Use')
print(g)
cat('\n','\n')

# modules size
g <- data_asv_s %>% 
  group_by(asv_group) %>% 
  summarise(n = n_distinct(asv_ID)) %>% 
  ggplot(aes(x=n)) +
  geom_histogram(binwidth = 1) +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14), strip.text.x = element_text(size=12)) +
  labs(x="Module Size", y="No. of Modules")
print(g)
cat('\n','\n')

# in how many land uses modules occur?
g <- data_asv_s %>% 
  group_by(asv_group) %>% 
  summarise(n = n_distinct(grid)) %>% 
  ggplot(aes(x=n)) +
  geom_histogram(binwidth = 1) +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14), strip.text.x = element_text(size=12)) +
  labs(x="No. of Land Uses", y="No. of Modules")
print(g)
cat('\n','\n')

###
cat('###', "Mantel-test for modules similarity",'\n')

# matrix of grid similarity in modules
grid_modules <- data_asv_s %>% 
  group_by(grid, asv_group) %>% 
  summarise(host_n = n_distinct(asv_ID)) %>% 
  spread(asv_group, host_n, fill = 0) %>% 
  column_to_rownames("grid") %>% 
  as.matrix()

grid_modules_dist <- as.matrix(1-vegdist(sqrt(grid_modules), method = "bray"))

# matrix of distances between grids
grid_attr <- data_asv_s %>% 
  distinct(grid) %>%
 left_join(grid_dist_avg, by="grid")

grid_dist <- expand_grid(id1=grid_attr$grid,id2=grid_attr$grid) %>% 
    left_join(grid_attr, by=c("id1"="grid")) %>% dplyr::rename(dist1=distance) %>% 
    left_join(grid_attr, by=c("id2"="grid")) %>% dplyr::rename(dist2=distance) %>% 
    mutate(distance = abs(dist1-dist2)) %>% 
    select(id1,id2,distance) %>% 
    spread(id2, distance) %>% 
    column_to_rownames("id1") %>% 
    as.matrix()

# mantel test
mantel_grid_modules <- ecodist::mantel(as.dist(grid_modules_dist) ~ as.dist(grid_dist))
# printing the results
mantel_grid_results <- tibble(r = mantel_grid_modules[1],
                              pvalue = mantel_grid_modules[3]) 
  print(knitr::kable(mantel_grid_results))
cat('\n','\n')

# plotting the correlation
grid_mudules_dist_m <- grid_modules_dist
grid_mudules_dist_m[lower.tri(grid_mudules_dist_m)] <- NA
diag(grid_mudules_dist_m) <- NA
grid_mudules_dist_m <- melt(grid_mudules_dist_m) %>% 
  filter(!is.na(value))

grid_dist_m <- melt(grid_dist) %>% dplyr::rename(distance=value)

g <- grid_mudules_dist_m %>% 
  left_join(grid_dist_m, by=c("Var1","Var2")) %>% 
ggplot(aes(y=value, x=distance)) +
  geom_point(alpha = 0.5, position=position_jitter(width=.1,height=.01)) +
  geom_smooth(method = "glm", se=T, method.args = list(family = "gaussian")) +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14), strip.text.x = element_text(size=12)) +
  labs(x = "Distance Between Grids [km]", y = "Modules Similarity [Bray-Curtis]")
print(g)
  cat('\n','\n')

  ###
cat('###', "Dendrogram",'\n')

# modules similarity between grids as tree
# calculate distance by UPGMA
phy <- as.dist(1-grid_modules_dist) %>%
  hclust(method = "average") 
print(plot(phy, main = NULL, xlab = NULL, ylab = "Modules Dissimilarity", ann=FALSE))
  cat('\n','\n')

}
```

# Microbes sharing
\
\

## Modules association to host species {.tabset}
\
\

### Modules exploration 
\
\

```{r}
# building a network
# including all hosts of both species

# converting to matrix
data_asv_mat <- data_asv %>% 
  mutate(reads = 1) %>% 
  spread(asv_ID, reads, fill = 0) %>% 
  column_to_rownames("host_ID") %>% 
  dplyr::select(starts_with("ASV")) %>% 
  as.matrix()

network_object <- infomapecology::create_monolayer_object(data_asv_mat, directed = FALSE, bipartite = TRUE, group_names = c("ASV", "Host"))

# finding modules
infomap_object <- run_infomap_monolayer(network_object,
                                        infomap_executable='Infomap',
                                        flow_model = 'undirected',
                                        two_level = TRUE,
                                        silent=TRUE, trials=100, seed=123)
cat('\n','\n')

# ASVs modules
modules_asv <- infomap_object$modules %>% 
  filter(node_group == "ASV") %>% 
  dplyr::rename(asv_group = module_level1, asv_ID = node_name) %>% 
  dplyr::select(asv_ID, asv_group)

# how many modules?
cat("How many modules: ", length(unique(modules_asv$asv_group)))
cat('\n','\n')

# modules size
g <- modules_asv %>% 
  group_by(asv_group) %>% 
  summarise(n = n_distinct(asv_ID)) %>% 
  ggplot(aes(x=n)) +
  geom_histogram(binwidth = 1) +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14), strip.text.x = element_text(size=12)) +
  labs(x="Module Size", y="No. of Modules")
print(g)
cat('\n','\n')
```

### Association to host species 
\
\

```{r}
# ASVs association to host species

# wide format
dat <- data_asv %>% 
  mutate(reads = 1) %>% 
  spread(asv_ID, reads, fill = 0)

# setting variables:
set.seed(123)
min_host <- 150
s <- 100 # number of shuffles
host_names <- unique(data_asv$host_species)
asv_group_names <- unique(modules_asv$asv_group)
asvs_host_shuff <- expand_grid(host_species=host_names,asv_group=asv_group_names) # dataframe for saving observed number of links in every host_species/asv_group (s cols)
#asvs_host_shuff_attr <- expand_grid(host_species=host_names,asv_group=asv_group_names) # dataframe for saving shuffled number of links in every grid/asv_group after shuffling grid attribute (s*s cols)

asv_group_host_association <- function(asv_host_dat) {
  # function for calculating the association between asv groups and grids
  # association = % of links of the asv_group in the grid out of all the asv_group links
  
  # total number of links in asv_group
  total_links <- asv_host_dat %>% 
    group_by(asv_group) %>% 
    summarise(total_n = sum(n, na.rm = TRUE))
    
  # calculating the association
  asv_host_assoc <- asv_host_dat %>% 
    left_join(total_links, by="asv_group") %>% 
    mutate(association = n/total_n)
   
  return(asv_host_assoc[c("host_species","asv_group","association")])
}


for(i in 1:s) {
  # randomly sampling hosts 
  dat_random <- dat %>% 
    group_by(host_species) %>% 
    slice_sample(n = min_host)
  
  # long format
  asvs <- dat_random %>%
    gather("asv_ID", "reads", starts_with("ASV")) %>% 
    filter(reads > 0) %>% 
    left_join(modules_asv, by="asv_ID")
  
  # observed number of links in every host/asv_group
  asvs_host <- asvs %>% count(host_species, asv_group)
  # calculating the association
  asv_host_assoc <- asv_group_host_association(asvs_host) %>% 
    rename_with(~paste("a",i,sep = "_"), association)
  asvs_host_shuff <- full_join(asvs_host_shuff, asv_host_assoc, by=c("host_species","asv_group")) # saving column
  
}

```

```{r}
# *mean* asv_group host association for the observed networks
#asvs_host_shuff %<>% mutate_all(~replace(., is.nan(.), 0))
obs_assoc <- rowMeans(asvs_host_shuff[,c(-1,-2)], na.rm = TRUE)
obs_assoc[is.nan(obs_assoc)] <- 0

# final table
host_assoc <- tibble(host_species = asvs_host_shuff$host_species,
                  asv_group = asvs_host_shuff$asv_group,
                  obs_association = obs_assoc)

# assigning association to host species 
asv_sign_assoc <- host_assoc %>% 
  filter(host_species=="Rattus rattus") %>% 
  mutate(association = ifelse(obs_association > 0.75, "Rattus rattus", ifelse(obs_association < 0.25, "Microgale brevicaudata", "General"))) %>% 
  select(asv_group, association)

# plotting the association
# plotting the association z-score
g <- asv_sign_assoc %>% 
  ggplot(aes(x = association, y = asv_group, fill = association)) +
  geom_tile() + 
  theme_classic()+
  theme(axis.text = element_text(size=14, color='black'), axis.text.x = element_text(angle = 45, hjust = 1))+
  labs(x="Association to Host Species",y="ASV Module")
print(g)
cat('\n','\n')

```

## Microbe sharing across land use {.tabset}
\
\

### Total sharing 
\
\
```{r}
# calculating microbe sharing

data_asv_shared <- data_asv %>% 
  left_join(modules_asv, by="asv_ID") %>% 
  left_join(asv_sign_assoc, by="asv_group")

# number of ASVs with association to host species
n_assoc <- data_asv_shared %>% distinct(asv_ID, association) %>% 
  count(association) 
  print(knitr::kable(n_assoc))
cat('\n','\n')

# finding shared (=occur in both host species) ASVs
shared_asv <- data_asv_shared %>% 
  group_by(asv_ID) %>% 
  summarise(shared_asv = n_distinct(host_species)) %>% 
  mutate(shared_asv = ifelse(shared_asv==2, 1, 0))

# total ASVs sharing %
cat("Total ASVs sharing:", nrow(shared_asv %>% filter(shared_asv==1))/nrow(shared_asv)*100,"%")
cat('\n','\n')

#add shared to the data
data_asv_shared %<>% left_join(shared_asv, by="asv_ID")

# porportion test
res <- prop.test(x = c(71, 38), n = c(295, 173))
cat("Proportion test- chi-square:",res$statistic, " p-value:", res$p.value)
cat('\n','\n')

# filter non-shared ASVs 
#data_asv_shared %<>% filter(shared_asv==1)

```

```{r}
# microbe sharing for specific host species

for(s in c("Rattus rattus","Microgale brevicaudata")) {

  cat("###", s, '\n','\n')

# does the host has a shared microbe?
# removing ASVs which are associated to the other host species, to the 'general' group, and to the village grid
shared_species <- data_asv_shared %>% 
  filter(host_species==s) %>% 
  filter(grid!="village") %>% 
  mutate(shared2 = ifelse(association==s,0,shared_asv)) %>% 
  #mutate(shared2 = ifelse(association=="General",0,shared2)) %>%
  group_by(host_ID, host_species, grid, dist_village) %>% 
  summarise(shared_host = max(shared2))

cat(table(shared_species$shared_host))
cat('\n','\n')

# logistic regression
logistic <- glm(shared_host ~ dist_village, data = shared_species, family = binomial)
reg_results <- as.data.frame(coef(summary(logistic)))
print(knitr::kable(reg_results))
cat('\n','\n')

g <- shared_species %>% 
ggplot(aes(y=shared_host, x=dist_village)) +
  geom_point(alpha = 0.5, position=position_jitter(width=.1,height=.01)) +
  geom_smooth(method = "glm", se=T, method.args = list(family = "binomial")) +
  theme_bw() +
  labs(x = "Distance from Village [km]", y = "Shared ASV")
print(g)
  cat('\n','\n')
  
}
```



```{r, include=FALSE}
# multilayer network (without interlayer edges)

# nodes metadata
nodes_host <- data_asv %>% distinct(host_ID) %>% rename(node_name = host_ID) %>%  mutate(type = "Host")
nodes_asv <- data_asv %>% distinct(asv_ID) %>% rename(node_name = asv_ID) %>%  mutate(type = "ASV")
nodes <- rbind(nodes_host, nodes_asv) %>% 
  rowid_to_column("node_id")

# layers metadata
layers <- data_asv %>% 
  distinct(host_species) %>% 
  rowid_to_column("layer_id")

# intra-layer edges (in extended format)
edges <- data_asv %>% 
  mutate(host_ID = as.character(host_ID)) %>% 
  select(host_species, host_ID, asv_ID) %>% 
  left_join(layers, by="host_species") %>% 
  left_join(nodes, by=c("asv_ID"="node_name")) %>% 
  rename(node_from=node_id) %>% 
  left_join(nodes, by=c("host_ID"="node_name")) %>% 
  rename(node_to=node_id) %>% 
  mutate(weight = 1) %>% 
  rename(layer_from=layer_id) %>% 
  mutate(layer_to = layer_from) %>% 
  select(layer_from, node_from, layer_to, node_to, weight) 

  a=c(1,1,2,1,0)
  edges=rbind(edges,a)

# creating a multilayer network
network_object <- infomapecology::create_multilayer_object(extended = edges, nodes = nodes, layers = layers)

infomap_object <- run_infomap_multilayer(network_object,
                                        infomap_executable='Infomap',
                                        flow_model = 'undirected',
                                        silent=TRUE, trials=100, seed=123)

modules <- infomap_object$modules
modules_asv <- infomap_object$modules %>% filter(type == "ASV")%>% rename(asv_group = module, asv_ID = node_name)
modules_host <- infomap_object$modules %>% filter(type == "Host")%>% rename(host_group = module, host_ID = node_name) %>% mutate(host_ID = as.numeric(host_ID))
cat('\n','\n')

data_asv %<>% left_join(modules_asv %>% dplyr::select(asv_ID,asv_group), by = "asv_ID") %>% 
  left_join(modules_host %>% dplyr::select(host_ID,host_group), by = "host_ID")

```
