---
title: "LBM_analysis"
output: 
  html_document:
    toc: true
    toc_float: true
    collapse: false
    number_sections: true
    code_folding: hide
    theme: united
    highlight: tango
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, results = "asis", message=FALSE, warning=FALSE, cache=TRUE, eval = TRUE, dev = c('png'), out.width = '100%', out.height='40%')
```


```{r load libraries, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(magrittr)
library(ggplot2)
library(vegan)
library(igraph)
library(blockmodels)
library(ape)
library(phangorn)
library(reshape2)
library(infomapecology)
library(aricode)

rm(list=ls())
```

```{r}
# reading ASV data
data_asv <- read_csv("data_processed/three_villages/data_asv_0.01.csv")
data_asv %<>% filter(host_species == "Rattus rattus" & village == "Mandena") 

# changing to 7 grids
full_grids <- read_csv("data_processed/three_villages/data_SM.csv") %>% dplyr::select(host_ID, grid)
data_asv %<>% dplyr::select(-grid) %>% left_join(full_grids, by="host_ID")

n_asv <- data_asv %>% 
   count(asv_ID) %>% 
   filter(n<10)
# # 
 #data_asv %<>% filter(asv_ID %in% n_asv$asv_ID)

data_asv_mat <- data_asv %>% 
  mutate(reads = 1) %>% 
  spread(asv_ID, reads, fill = 0) %>% 
  column_to_rownames("host_ID") %>% 
  dplyr::select(starts_with("ASV")) %>% 
  as.matrix()
```

# The Model {.tabset} 
I analyze **un-weighted bipartite** network of ASVs and individual hosts.\
**Here the analysis includes only Rattus from Mandena**\
I use Latent Block Model (LBM) from the *blockmodels* package.\
\

```{r}
# Latent Block Model
model <- blockmodels::BM_bernoulli(data_asv_mat, membership_type = "LBM", verbosity=0, plotting="")
#model <- blockmodels::BM_bernoulli(data_asv_mat, membership_type = "LBM", exploration_direction = c(10,10), exploration_factor = 10, explore_min = 1, explore_max = 100)
model$estimate()

# the best model
b <- which.max(model$ICL)
memb_host <- as.data.frame(model$memberships[[b]]$Z1)
memb_asv <- as.data.frame(model$memberships[[b]]$Z2)

# attaching hosts and ASVs IDs
host_names <- rownames(data_asv_mat)
asv_names <- colnames(data_asv_mat)
memb_host %<>% mutate(host_group = max.col(.,'first')) %>% mutate(host_ID = as.numeric(host_names))
memb_asv %<>% mutate(asv_group = max.col(.,'first')) %>% mutate(asv_ID = asv_names)

# matching host group to ASV group
data_asv %<>% left_join(memb_host %>% dplyr::select(host_ID,host_group), by = "host_ID") %>% 
  left_join(memb_asv %>% dplyr::select(asv_ID,asv_group), by = "asv_ID")
#write_csv(data_asv, "data_processed/three_villages/data_asv_0.01_rattus_mandena_groups.csv")

cat("## Hosts", '\n')
print(knitr::kable(head(memb_host)))
cat('\n','\n')

cat("## ASVs", '\n')
print(knitr::kable(head(memb_asv)))
cat('\n','\n')

cat("## Plot", '\n')
print(model$memberships[[b]]$plot())
cat('\n','\n')
```


# Exploring Clustering associations {.tabset}
\
\

## ASV-Host association - Model parameters
values = the probabilities of connections or edges between nodes in different blocks as found by the model.\
represent the probabilities of an edge existing between a pair of blocks.\
\
```{r}
# pi parameters by the model
p <- as.data.frame(model$model_parameters[[b]]$pi) %>% 
  rownames_to_column("host_group") %>% 
  gather("asv_group", "pi", starts_with("V")) %>% 
  mutate(asv_group = as.numeric(gsub(".*?([0-9]+).*", "\\1", asv_group)))

# plotting
g <- p %>% 
  ggplot(aes(x = host_group, y = asv_group, fill = pi, label=round(pi,3))) +
  geom_tile() + 
  geom_text()+
  theme_classic()+
  scale_fill_distiller(palette = "Spectral", limits = c(0,1)) +  
theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14)) +
  labs(x="Host Group", y="ASVs Group")
print(g)
cat('\n','\n')
```


# Exploring ASVs groups {.tabset}
\
\

## Relative abundance
\
\
```{r}
# ASVs relative abundance across clustering groups
avg_abundance <- data_asv %>% 
  group_by(asv_ID) %>% 
  summarise(rel_abu = mean(reads)) %>% 
  bind_cols(asv_group = memb_asv$asv_group)
# plotting
g <- avg_abundance %>% 
  ggplot(aes(x=as.factor(asv_group), y=rel_abu)) + 
  geom_boxplot(fill="slateblue") + 
  theme_bw()
print(g)
cat('\n','\n')
```

## Occurrence
\
\
```{r}
# ASVs occurrence across clustering groups
asv_occur <- data_asv %>% 
  group_by(asv_ID) %>% 
  summarise(occur = n_distinct(host_ID)) %>% 
  mutate(occur = occur/length(unique(data_asv$host_ID))) %>% 
  bind_cols(asv_group = memb_asv$asv_group)
# plotting
g <- asv_occur %>% 
  ggplot(aes(x=as.factor(asv_group), y=occur)) + 
  geom_boxplot(fill="slateblue") + 
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14)) +
  labs(x="ASV Group", y="Occurence")
print(g)
cat('\n','\n')
```

## ASVs Season
\
\
```{r}
# ASVs seasons across clustering groups
g <- data_asv %>% 
  group_by(asv_group) %>% 
  count(season) %>% 
  ggplot(aes(x=as.factor(asv_group), y=n, fill=as.factor(season))) + 
  geom_bar(position="fill", stat="identity") + 
  theme_bw()
print(g)
cat('\n','\n')
```

Shuffling season and calculating Normalized Mutual Information (NMI)\

```{r}
# wide format
dat <- data_asv %>% 
  mutate(reads = 1) %>% 
  dplyr::select(-host_group,-asv_group) %>% 
  spread(asv_ID, reads, fill = 0)

min_grid <- dat %>% count(season) # number of individuals in each season
s <- 100 # number of shuffles
nmi_obs_shuff <- vector(length = s)
nmi_shuff <- vector(length = s*s)
asvs_grid_shuff <- data.frame(season=NA,asv_group=NA)
asvs_n_shuff <- data.frame(asv_group=NA)

for(i in seq(0,s*(s-1), by = s)) {
  # randomly sampling hosts 
  dat_random <- dat %>% 
    group_by(season) %>% 
    slice_sample(n = min(min_grid$n))
  
  # long format
  asvs <- dat_random %>%
    gather("asv_ID", "reads", starts_with("ASV")) %>% 
    filter(reads > 0) %>% 
    left_join(memb_asv %>% dplyr::select(asv_ID,asv_group), by="asv_ID")
  
  asvs_grid <- asvs %>% count(season, asv_group)
  asvs_grid_shuff <- full_join(asvs_grid_shuff, asvs_grid, by=c("season","asv_group"))
  
  asvs_n <- asvs %>% group_by(asv_group) %>% summarise(n = n_distinct(asv_ID))
  asvs_n_shuff <- full_join(asvs_n_shuff, asvs_n, by="asv_group")
  
  nmi_obs_shuff[(i/s)+1] <- aricode::NMI(asvs$season, asvs$asv_group)
  
  for(j in 1:s) {
    # shuffling the season attribute
    asvs_shuff <- sample(asvs$season,nrow(asvs))
    
    # calculating nmi
    nmi_shuff[i+j] <- aricode::NMI(asvs_shuff, asvs$asv_group)
  }
}

# plotting
nmi_obs_mean <- mean(nmi_obs_shuff)
nmi_obs_sd <- sd(nmi_obs_shuff)
p <- length(nmi_shuff[nmi_shuff>nmi_obs_mean]) / length(nmi_shuff)
g <- as.data.frame(nmi_shuff) %>% 
  ggplot(aes(nmi_shuff)) + 
  geom_histogram() + 
  theme_bw() +
  geom_vline(xintercept = c(nmi_obs_mean-nmi_obs_sd, nmi_obs_mean, nmi_obs_mean+nmi_obs_sd), linetype='dashed', color=c("blue","red","blue")) +
  annotate("text", x=max(nmi_shuff), y=1000, label=paste('p-value =',p))
print(g)
cat('\n','\n')

asvs_grid_shuff <- asvs_grid_shuff[-1,]
asvs_grid_n <- rowMeans(asvs_grid_shuff[,-(1:2)], na.rm = TRUE)
asvs_n_shuff <- asvs_n_shuff[-1,]
asvs_total_n <- tibble(asv_group = unique(asvs_n_shuff$asv_group),
                       n_total = rowMeans(asvs_n_shuff[,-1], na.rm = TRUE))


asvs_grid_shuff_n <- asvs_grid_shuff %>% 
  dplyr::select(season, asv_group) %>% 
  mutate(n = asvs_grid_n) %>% 
  left_join(asvs_total_n, by="asv_group") %>% 
  mutate(max_density = n_total * min(min_grid$n)) %>% 
  mutate(density = n/max_density)

cat("Density",'\n')
g <- asvs_grid_shuff_n %>% 
  ggplot(aes(x = season, y = asv_group, fill = density, label=round(density,2))) +
  geom_tile() + 
  geom_text()+
  theme_classic()+
  scale_fill_distiller(palette = "Spectral", limits = c(0,1)) +  
  theme(axis.text = element_text(size=12, color='black'), axis.text.x = element_text(angle = 45, hjust = 1))
print(g)
cat('\n','\n')
```

## Phylogenetic distance
I calculate the *median* phylogenetic distance within each group (Intra) and between groups (Inter).\
Then I calculate log(intra/inter) between all groups.\
Negative values = group x is closer to itself than to other groups.\
\
```{r}
# reading phylogenetic tree
phylo_tree <- readRDS("output/phylogenetic_tree/phylo_tree_0.01.rds")
tree <- phylo_tree$tree
rooted_tree <- midpoint(tree) # rooting the tree
# prune the tree
included_asvs <- unique(data_asv$asv_ID)
unincluded <- rooted_tree$tip.label[!rooted_tree$tip.label %in% included_asvs]
pruned_tree <- dendextend::prune(rooted_tree, unincluded)

# extracting phylogenetic distances from the tree
phylo_dist <- ape::cophenetic.phylo(pruned_tree)
phylo_dist_m <- melt(phylo_dist)
phylo_cluster <- phylo_dist_m %>% 
  left_join(memb_asv %>% dplyr::select(asv_ID,asv_group), by=c("Var1"="asv_ID")) %>% 
  left_join(memb_asv %>% dplyr::select(asv_ID,asv_group), by=c("Var2"="asv_ID")) %>% 
  filter(Var1 != Var2)

min_asv <- phylo_cluster %>% count(asv_group.x) 
phylo_calc_shuff <- tibble()

for(i in 1:100) {
  # randomly sampling asvs 
  dat_random <- phylo_cluster %>% 
    group_by(asv_group.x) %>% 
    slice_sample(n = min(min_asv$n))
  
  # median phylogenetic distance between microbe groups
  avg_phylo_dist <- dat_random %>% 
    group_by(asv_group.x, asv_group.y) %>% 
    summarise(median_dist = median(value)) 
  
  # intra- group phylogenetic distance
  intra_dist <- avg_phylo_dist %>% filter(asv_group.x==asv_group.y) %>% 
    dplyr::select(-asv_group.y) %>% rename("intra_median_dist" = "median_dist")
  
  # compare intra to inter distance
  # positive values = group x is closer to group y than to group x
  phylo_calc <- avg_phylo_dist %>% 
    left_join(intra_dist, by="asv_group.x") %>% 
    filter(asv_group.x != asv_group.y) %>% 
    mutate(calc = log(intra_median_dist/median_dist)) %>% 
    dplyr::select(-median_dist,-intra_median_dist)
  
  phylo_calc_shuff <- bind_rows(phylo_calc_shuff, phylo_calc)
  
}

# plotting
g <- phylo_calc_shuff %>% 
  ggplot(aes(x = as.factor(asv_group.x), y = calc)) +
  geom_boxplot(fill="slateblue") + 
  theme_classic()+
  geom_hline(yintercept = 0, linetype='dashed', color="red") +
  #scale_fill_distiller(palette = "Spectral", limits = c(-2,2))+  
  theme(axis.text = element_text(size=12, color='black'))
print(g)
cat('\n','\n')
```

## ASVs Taxonomy
Proportion of taxa (minimum 2 ASVs per taxa)\
\
```{r}
# reading ASVs taxonomy
asv_taxa <- read_delim("data_raw/ASVs_all_merged_taxonomy.tsv")
memb_asv_taxa <- memb_asv %>%  left_join(asv_taxa, by=c("asv_ID"="ASV"))


g <- memb_asv_taxa %>% 
  group_by(asv_group) %>% 
  count(Family) %>% 
  filter(n>=2) %>% 
  ggplot(aes(x=as.factor(asv_group), y=n, fill=Family)) + 
  geom_bar(position="fill", stat="identity") + 
  theme_bw()+
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14)) +
  labs(x="ASV Group", y="No. of ASVs [%]")
print(g)
cat('\n')

g <- memb_asv_taxa %>% 
  group_by(asv_group) %>% 
  count(Phylum) %>% 
  # filter(n>=3) %>% 
  ggplot(aes(x=as.factor(asv_group), y=n, fill=Phylum)) + 
  geom_bar(position="fill", stat="identity") + 
  theme_bw()+
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14)) +
  labs(x="ASV Group", y="No. of ASVs [%]")
print(g)
cat('\n','\n')
```

# ASVs Land-Use {.tabset}
\
\

```{r}
# wide format
dat <- data_asv %>% 
  mutate(reads = 1) %>% 
  dplyr::select(-host_group,-asv_group) %>% 
  spread(asv_ID, reads, fill = 0)

# setting variables:
min_grid <- dat %>% count(grid) # number of individuals in each grid
s <- 100 # number of shuffles
grid_names <- unique(data_asv$grid)
asv_group_names <- unique(data_asv$asv_group)
nmi_obs_shuff <- vector(length = s) # s observed NMI values
nmi_shuff <- vector(length = s*s) # s*s shuffled (i.e., grid host attribute) NMI values
asvs_grid_shuff <- expand_grid(grid=grid_names,asv_group=asv_group_names) # dataframe for saving observed number of links in every grid/asv_group (s cols)
asvs_grid_shuff_attr <- expand_grid(grid=grid_names,asv_group=asv_group_names) # dataframe for saving shuffled number of links in every grid/asv_group after shuffling grid attribute (s*s cols)

asv_group_grid_association <- function(asv_grid_dat) {
  # function for calculating the association between asv groups and grids
  # association = % of links of the asv_group in the grid out of all the asv_group links
  
  
  # total number of links in asv_group
  total_links <- asv_grid_dat %>% 
    group_by(asv_group) %>% 
    summarise(total_n = sum(n, na.rm = TRUE))
    
  # calculating the association
  asv_grid_assoc <- asv_grid_dat %>% 
    left_join(total_links, by="asv_group") %>% 
    mutate(association = n/total_n)
   
  return(asv_grid_assoc[c("grid","asv_group","association")])
}


for(i in seq(0,s*(s-1), by = s)) {
  # randomly sampling hosts 
  dat_random <- dat %>% 
    group_by(grid) %>% 
    slice_sample(n = min(min_grid$n))
  
  # long format
  asvs <- dat_random %>%
    gather("asv_ID", "reads", starts_with("ASV")) %>% 
    filter(reads > 0) %>% 
    left_join(memb_asv %>% dplyr::select(asv_ID,asv_group), by="asv_ID")
  
  # observed number of links in every grid/asv_group
  asvs_grid <- asvs %>% count(grid, asv_group)
  # calculating the association
  asv_grid_assoc <- asv_group_grid_association(asvs_grid) %>% 
    rename_with(~paste("a",i,sep = "_"), association)
  asvs_grid_shuff <- full_join(asvs_grid_shuff, asv_grid_assoc, by=c("grid","asv_group")) # saving column
  
  # saving observed NMI value
  nmi_obs_shuff[(i/s)+1] <- aricode::NMI(asvs$grid, asvs$asv_group)
  
  for(j in 1:s) {
    # shuffling the grid attribute
    asvs_shuff <- sample(asvs$grid,nrow(asvs))
    asvs$grid <- asvs_shuff
    
    # calculating nmi
    nmi_shuff[i+j] <- aricode::NMI(asvs_shuff, asvs$asv_group)
    
    # shuffled number of links in every grid/asv_group
    asvs_grid_attr <- asvs %>% count(grid, asv_group) 
    # calculating the association
    asv_grid_attr_assoc <- asv_group_grid_association(asvs_grid_attr) %>% 
    rename_with(~paste("a",(i+j),sep = "_"), association)
    asvs_grid_shuff_attr <- full_join(asvs_grid_shuff_attr, asv_grid_attr_assoc, by=c("grid","asv_group")) # saving column

  }
}

```

## Z-scores
\
```{r}


# *mean* asv_group grid association for the observed networks
asvs_grid_shuff %<>% mutate_all(~replace(., is.na(.), 0))
obs_assoc <- rowMeans(asvs_grid_shuff[,c(-1,-2)])
  

# *mean* asv_group grid association for the shuffled networks
asvs_grid_shuff_attr %<>% mutate_all(~replace(., is.na(.), 0))
shuff_assoc <- rowMeans(asvs_grid_shuff_attr[,c(-1,-2)])
# standard deviation of the shuffled networks
shuff_sd <- apply(asvs_grid_shuff_attr, 1, sd, na.rm=TRUE)

# calculating Z-scores:

# final table
z_scores <- tibble(grid = asvs_grid_shuff$grid,
                  asv_group = asvs_grid_shuff$asv_group,
                  obs_association = obs_assoc,
                  shuff_association = shuff_assoc,
                  shuff_sd = shuff_sd) %>% 
  mutate(z_score = (obs_association-shuff_association)/shuff_sd)


# plotting the association

# plotting the association z-score
g <- z_scores %>% 
    mutate(grid = factor(grid, levels = c("semi-intact_forest","secondary_forest","brushy_regrowth","agriculture","flooded_rice","agroforest","village"))) %>% 
  mutate(z_score_d = ifelse(z_score>1.96,1,ifelse(z_score< (-1.96),-1,0))) %>% 
  ggplot(aes(x = grid, y = asv_group, fill = z_score_d, label=round(z_score,2))) +
  geom_tile(color = "black") + 
  geom_text()+
  theme_classic()+
  scale_fill_gradient2(low = "#d62828",
                       mid = "#edf2f4",
                       high = "#2b9348") +
theme(axis.text = element_text(size=12, color='black'), axis.text.x = element_text(angle = 45, hjust = 1), title = element_text(size = 14),legend.position = "none")+
  labs(x="Land Use", y="ASV Group")
print(g)
cat('\n','\n')

g <- z_scores %>% 
      mutate(grid = factor(grid, levels = c("semi-intact_forest","secondary_forest","brushy_regrowth","agriculture","flooded_rice","agroforest","village"))) %>% 
  ggplot(aes(x = grid, y = asv_group, fill = obs_association, label=round(obs_association,2))) +
  geom_tile() + 
  geom_text()+
  theme_classic()+
  scale_fill_distiller(palette = "Spectral", limits = c(0,1)) +  
  theme(axis.text = element_text(size=12, color='black'), axis.text.x = element_text(angle = 45, hjust = 1), title = element_text(size = 14))+
  labs(x="Land Use", y="ASV Group")
print(g)
cat('\n','\n')
```

## % of ASVs
\
```{r}
# ASVs grids across clustering groups
g <- data_asv %>% 
  group_by(asv_group) %>% 
  count(grid) %>% 
  ggplot(aes(x=as.factor(asv_group), y=n, fill=grid)) + 
  geom_bar(position="fill", stat="identity") + 
  theme_bw()
print(g)
cat('\n','\n')
```

## NMI
\
```{r}
nmi_obs_mean <- mean(nmi_obs_shuff)
p <- length(nmi_shuff[nmi_shuff>nmi_obs_mean]) / length(nmi_shuff)
g <- as.data.frame(nmi_shuff) %>% 
  ggplot(aes(nmi_shuff)) + 
  geom_histogram() + 
  theme_bw() +
  geom_vline(xintercept = nmi_obs_mean, linetype='dashed', color="red") +
  xlab("NMI")+
  annotate("text", x=max(nmi_shuff), y=1000, label=paste('p-value =',p))
print(g)
cat('\n','\n')
```

## Beta-Diversity
asv_group similarity [Jaccard] across grids\
value = average similarity of the grid to all other grids\
\
```{r}
d <- data_asv %>% 
  distinct(grid, asv_group, asv_ID)

mean_sim_all <- NULL
for(l in asv_group_names){
  d_group <- d %>% filter(asv_group == l) %>% 
  mutate(reads = 1) %>% 
  dplyr::select(-asv_group) %>% 
  spread(asv_ID, reads, fill = 0) %>% 
  column_to_rownames("grid") %>% 
    as.matrix()
  
  sim <- as.matrix(1-vegdist(d_group, method = "jaccard"))
  diag(sim) <- NA
  mean_sim <- tibble(grid = rownames(sim),
                     asv_group = l,
    similarity = rowMeans(sim, na.rm = TRUE))
  
  mean_sim_all %<>% bind_rows(mean_sim)
}

# plotting
g <- mean_sim_all %>% 
  ggplot(aes(x = grid, y = asv_group, fill = similarity, label=round(similarity,2))) +
  geom_tile() + 
  geom_text()+
  theme_classic()+
  scale_fill_distiller(palette = "RdYlGn", direction = -1, limits = c(0,1))+
  theme(axis.text = element_text(size=12, color='black'), axis.text.x = element_text(angle = 45, hjust = 1))
print(g)
cat('\n','\n')
```


# Exploring Host groups {.tabset}
\
\

## ASVs Richness
\
\
```{r}
# ASVs average (per capita) richness across clustering groups
asv_richness <- data_asv %>% 
  group_by(host_group, host_ID) %>% 
  summarise(asv_richness = n_distinct(asv_ID)) 

# plotting
g <- asv_richness %>% 
  ggplot(aes(x=as.factor(host_group), y=asv_richness)) + 
  geom_boxplot(fill="slateblue") + 
  theme_bw()+
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14)) +
  labs(x="Host Group", y="Average ASVs Richness")
print(g)
cat('\n','\n')
```

## ASVs Group Richness
\
\
```{r}
# ASVs average (per capita) group richness across clustering groups
asv_group_richness <- data_asv %>% 
  group_by(host_group, host_ID) %>% 
  summarise(asv_group_richness = n_distinct(asv_group)) 

# plotting
g <- asv_group_richness %>% 
  ggplot(aes(x=as.factor(host_group), y=asv_group_richness)) + 
  geom_boxplot(fill="slateblue") + 
  theme_bw()
print(g)
cat('\n','\n')
```

## Host Mass
\
\
```{r}
# Host average mass across clustering groups
host_mass <- data_asv %>% 
  group_by(host_group) %>% 
  summarise(host_mass = mean(mass)) 

# plotting
g <- data_asv %>% 
  ggplot(aes(x=as.factor(host_group), y=mass)) + 
  geom_boxplot(fill="slateblue") + 
  theme_bw()
print(g)
cat('\n','\n')
```

## Host Sex Ratio
\
\
```{r}
# Host sex ratio across clustering groups
g <- data_asv %>% 
  group_by(host_group) %>% 
  count(sex) %>% 
  ggplot(aes(x=as.factor(host_group), y=n, fill=sex)) + 
  geom_bar(position="fill", stat="identity") + 
  theme_bw()
print(g)
cat('\n','\n')
```

## Host Land-Use
\
\
```{r}
# Host grids across clustering groups
g <- data_asv %>% 
  group_by(host_group) %>% 
  count(grid) %>% 
  ggplot(aes(x=as.factor(host_group), y=n, fill=grid)) + 
  geom_bar(position="fill", stat="identity") + 
  theme_bw()+
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14)) +
  labs(x="Host Group", y="No. of Hosts [%]")
print(g)
cat('\n')
```

Shuffling land-use and calculating Normalized Mutual Information (NMI)\

```{r}
hosts <- data_asv %>% distinct(host_ID, grid, host_group) 

nmi_obs <- aricode::NMI(hosts$grid, hosts$host_group)
nmi_shuff <- vector(length = 1000)
for(i in 1:1000) {
  # shuffling the grid attribute
  hosts_shuff <- hosts %>% 
    mutate(grid = sample(grid,nrow(hosts)))
  
  # calculating nmi
  nmi_shuff[i] <- aricode::NMI(hosts_shuff$grid, hosts_shuff$host_group)
}

# plotting
p <- length(nmi_shuff[nmi_shuff>nmi_obs]) / length(nmi_shuff)
g <- as.data.frame(nmi_shuff) %>% 
  ggplot(aes(nmi_shuff)) + 
  geom_histogram() + 
  theme_bw() +
  geom_vline(xintercept = nmi_obs, linetype='dashed', color="red") +
  theme(axis.text = element_text(size = 14, color = 'black'), title = element_text(size = 20)) +
  labs(x="NMI", y="No. of Shuffled Networks")+
  annotate("text", x=max(nmi_shuff), y=100, label=paste('p-value =',p))
ggsave("nmi_host_grid.pdf", path = "LBM_analysis_new_files", device = "pdf")
print(g)
cat('\n','\n')
```

```{r}
# wide format
dat <- data_asv %>% 
  distinct(host_ID, grid, host_group)

# setting variables:
min_grid <- dat %>% count(grid) # number of individuals in each grid
s <- 100 # number of shuffles
grid_names <- unique(dat$grid)
host_group_names <- unique(dat$host_group)
nmi_obs_shuff <- vector(length = s) # s observed NMI values
nmi_shuff <- vector(length = s*s) # s*s shuffled (i.e., grid host attribute) NMI values

for(i in seq(0,s*(s-1), by = s)) {
  # randomly sampling hosts 
  dat_random <- dat %>% 
    group_by(grid) %>% 
    slice_sample(n = min(min_grid$n))
  
  # saving observed NMI value
  nmi_obs_shuff[(i/s)+1] <- aricode::NMI(dat_random$grid, dat_random$host_group)
  
  for(j in 1:s) {
    # shuffling the grid attribute
    dat_shuff <- sample(dat_random$grid,nrow(dat_random))
    dat_random$grid <- dat_shuff
    
    # calculating nmi
    nmi_shuff[i+j] <- aricode::NMI(dat_shuff, dat_random$host_group)
  }
}

# plotting
nmi_obs <- mean(nmi_obs_shuff)
p <- length(nmi_shuff[nmi_shuff>nmi_obs]) / length(nmi_shuff)
g <- as.data.frame(nmi_shuff) %>% 
  ggplot(aes(nmi_shuff)) + 
  geom_histogram() + 
  theme_bw() +
  geom_vline(xintercept = nmi_obs, linetype='dashed', color="red") +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14)) +
  labs(x="NMI", y="No. of Shuffled Networks")+
  annotate("text", x=nmi_obs, y=1000, label=paste('p-value =',p))
print(g)
cat('\n','\n')

```


## Host Season
\
\
```{r}
# Host season across clustering groups
g <- data_asv %>% 
  group_by(host_group) %>% 
  count(season) %>% 
  ggplot(aes(x=as.factor(host_group), y=n, fill=as.factor(season))) + 
  geom_bar(position="fill", stat="identity") + 
  theme_bw()
print(g)
cat('\n','\n')
```

```{r}
# what explain host groups?

host_group_mat <- memb_host %>%
  dplyr::select(-host_group) %>% 
  column_to_rownames("host_ID") %>% 
  as.matrix()

host_richness <- data_asv %>% count(host_ID)
host_attr <- data_asv %>%
  distinct(host_ID, grid, season, mass, sex) %>% 
  left_join(host_richness, by="host_ID")
  
# PERMANOVA
perm_ratus <- as.data.frame(adonis2(formula = host_group_mat ~ as.factor(grid)+as.factor(season)+mass+as.factor(sex)+n, data = host_attr,  method = "bray"))
perm_ratus
```

```{r}
# what explain asv groups?

asv_group_mat <- memb_asv %>%
  dplyr::select(-asv_group) %>% 
  column_to_rownames("asv_ID") %>% 
  as.matrix()

asv_pre <- data_asv %>% count(asv_ID)

  
# PERMANOVA
perm_ratus <- as.data.frame(adonis2(formula = asv_group_mat ~ n, data = asv_pre,  method = "bray"))
perm_ratus
```


