---
title: "Data exploration"
output: 
  html_document:
    toc: true
    toc_float: true
    collapse: false
    number_sections: true
    code_folding: hide
    theme: united
    highlight: tango
  
    
  
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = FALSE, results = "asis", message=FALSE, warning=FALSE, cache=TRUE, eval = TRUE, dev = c('png','pdf'), out.width = '100%', out.height='40%')
```


```{r load libraries, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(magrittr)
library(ggplot2)
library(vegan)
library(lme4)
library(multcomp) # post-hoc for glmm
library(fitdistrplus) # post-hoc for dispersion
library(reshape2)
library(MuMIn) # for AIC
library(RVAideMemoire) # for pairwise PERMANOVA
library(phangorn) # phylogenetic tree
library(GUniFrac) # UniFrac
library(picante) # phylogentic diversity

rm(list=ls())
```

```{r}
# 1. filtering *small-mammals* data

data_mammals <- read_csv("data_raw/Terrestrial_Mammals.csv")

# reading ASVs raw data
data_asv <- read_csv("data_raw/merged_full_sample_table.csv") %>%  filter(sample_type == "SAMPLE") %>% 
   mutate(across(starts_with("ASV"),~ ./unfiltered_reads)) 


# extracting samples IDs (taking only numbers due to unmatching formats)
data_mammals %<>% mutate(host_ID = as.numeric(gsub(".*?([0-9]+).*", "\\1", animal_id)))
data_asv %<>% mutate(host_ID = as.numeric(gsub(".*?([0-9]+).*", "\\1", Sample_Name)))

# matching small-mammals (SM) IDs in the two data files and taking only SM with microbes data
data_sm <- semi_join(data_mammals, data_asv, by="host_ID") %>% 
  dplyr::select(host_ID, field_identification, village, habitat_type, season, sex, mass) %>%
  dplyr::rename(host_species = field_identification, grid = habitat_type) %>% 
  mutate(season = factor(season, levels = c("1","2","3"))) %>% 
   mutate(grid = factor(grid, levels = c("semi-intact_forest","secondary_forest","brushy_regrowth","agriculture","flooded_rice","agroforest","village")))
 # write_csv(dat_long, "data_processed/three_villages/data_asv_0.001.csv")
```
semi-intact forest = primary forest\
secondary forest\
brushy regrowth = savoka\
agriculture = sugar cane, mixed agriculture, coffee, banana\
agroforest = vanilla\
flooded rice = rice\
village\
\
Merging similar land uses together:\
1. semi-intact forest\
2. agriculture = secondary_forest + brushy_regrowth + agriculture + agroforest\
3. flooded_rice\
4. village\

```{r, include=FALSE}
data_sm %<>% mutate_at(vars(grid), funs(case_when(grid=="secondary_forest"|grid=="brushy_regrowth"|grid=="agriculture"|grid=="agroforest"  ~ "agriculture", TRUE ~ .))) %>% 
  mutate(grid = factor(grid, levels = c("semi-intact_forest", "agriculture", "flooded_rice", "village")))
```


# Small-mammal exploration {.tabset}
Exploration of small mammals abundance and richness across land use\
\

## Abundance
```{r}
# total SM abundance by grid
data_sm %>% count(village, grid) %>% 
  ggplot(aes(x=grid, y=n, fill = village)) +
  geom_bar(position="dodge", stat="identity") +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black', angle = 45, vjust = 0.5, hjust=1), title = element_text(size = 14)) +
  labs(x="Land-use", y="Total small-mammals abundance") 
```

## Richness
```{r}
# total SM richness by grid
data_sm %>% group_by(grid, village) %>% 
  summarise(richness=n_distinct(host_species)) %>% 
  ggplot(aes(fill=village, x=grid, y=richness)) +
  geom_bar(position="dodge", stat="identity") +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black', angle = 45, vjust = 0.5, hjust=1), title = element_text(size = 14)) +
  scale_y_continuous(breaks = seq(0, 10, by = 1)) +
  labs(x="Land-use", y="Small-mammals richness") 
```

## Abundance by sp.
```{r}
# SM abundance by grid for each species separately
data_sm %>% count(village, grid, host_species) %>% 
  ggplot(aes(fill=village, x=grid, y=n)) +
  geom_bar(position="dodge", stat="identity") +
  facet_wrap(~host_species) +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black', angle = 90, vjust = 0.5, hjust=1), title = element_text(size = 14)) +
  labs(x="Land-use", y="Small-mammals abundance") 
```

## Common sp.
```{r}
# SM abundance by grid for each species separately - common species
data_sm %>% filter(host_species=="Rattus rattus" | host_species=="Microgale brevicaudata" | host_species=="Mus musculus") %>% 
  count(village, grid, host_species) %>% 
  ggplot(aes(fill=village, x=grid, y=n)) +
  geom_bar(position="dodge", stat="identity") +
  facet_wrap(~host_species) +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black', angle = 90, vjust = 0.5, hjust=1), title = element_text(size = 14)) +
  labs(x="Land-use", y="Small-mammals abundance") 

data_sm %>% filter(host_species=="Rattus rattus" | host_species=="Microgale brevicaudata" | host_species=="Mus musculus") %>% 
  count(village, host_species) %>% 
   knitr::kable()
```

# Microbiome data exploration

```{r,  include=FALSE}
# exploring and comparing samples with both buffers
data_double <- data_asv %>% 
  group_by(host_ID) %>% filter(n()>1) %>% 
  mutate(buffer = ifelse(grepl("D",Sample_Name),1,2)) %>% 
  arrange(host_ID)
id_double <- unique(data_double$host_ID)
for(d in id_double) {
  comp <- data_double %>% filter(host_ID==d) %>% dplyr::select(contains("ASV"))
  sim <- 1-vegdist(comp, "jaccard")
  #print(sim)
}
```

```{r}
# filtering the data - only the common host species
data_asv_f <- data_asv %>% 
  filter(!(grepl("D",Sample_Name) & host_ID %in% id_double)) %>% 
  dplyr::select(host_ID, unfiltered_reads, contains("ASV"))

dat <- left_join(data_sm, data_asv_f, by="host_ID") %>% 
  filter(host_species=="Rattus rattus" | host_species=="Microgale brevicaudata" | host_species=="Mus musculus") %>% 
  filter(!(is.na(unfiltered_reads)))
```

## Filtering by abundance {.tabset}

Filtering ASVs by relative abundance\
ASVs with relative reads count of less than x% within a sample are filtered out

### Proportion left

```{r}
dat_long <- dat %>% gather("asv_ID", "reads", starts_with("ASV")) %>% filter(reads > 0) # transforming to long format
n_asv_sp <- dat_long %>% group_by(host_species) %>% summarise(n = n_distinct(asv_ID)) # total number of ASVs before filtering
sensitivity_abundance <- NULL
for (t in seq(0,0.03,0.001)) {
  # t = cutoff for excluding ASV/sample combinations that accounted for <X% of total reads
  
  dat_long %<>% filter(reads >= t)
  asv_left <- dat_long %>% group_by(host_species) %>% summarise(n = n_distinct(asv_ID)) %>% 
    bind_cols(total_n = n_asv_sp$n) %>% 
    mutate(proportion = n/total_n, t=t*100)
  sensitivity_abundance <- bind_rows(sensitivity_abundance, asv_left)
}

# plotting
sensitivity_abundance %>% 
  ggplot(aes(t, proportion, color=host_species)) +
  geom_line() +
  geom_point() +
  geom_vline(xintercept = c(0.1, 0.5, 1), linetype='dashed') +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14)) +
  labs(x="Abundance cutoff [%]", y="ASV richness proportion left")
cat('\n')

cat("Number of ASVs \n")
sensitivity_abundance %>% 
  dplyr::select(t,host_species,n) %>% 
  spread(host_species,n) %>% 
  filter(t %in% c(0,0.1,0.2,0.3,0.4,0.5,1,2)) %>% 
  knitr::kable()
```

### Proportion left by Grid

```{r}
dat_long <- dat %>% gather("asv_ID", "reads", starts_with("ASV")) %>% filter(reads > 0) # transforming to long format
n_asv_sp <- dat_long %>% group_by(grid, host_species) %>% summarise(n = n_distinct(asv_ID)) # total number of ASVs before filtering
sensitivity_abundance <- NULL
for (t in seq(0,0.03,0.001)) {
  # t = cutoff for excluding ASV/sample combinations that accounted for <X% of total reads
  
  dat_long %<>% filter(reads >= t)
  asv_left <- dat_long %>% group_by(grid, host_species) %>% summarise(n = n_distinct(asv_ID)) %>% 
    bind_cols(total_n = n_asv_sp$n) %>% 
    mutate(proportion = n/total_n, t=t*100)
  sensitivity_abundance <- bind_rows(sensitivity_abundance, asv_left)
}

# plotting
sensitivity_abundance %>% 
  ggplot(aes(t, proportion, color=host_species)) +
  geom_line() +
  geom_point() +
  facet_wrap(~grid) +
  geom_vline(xintercept = c(0.1, 0.5, 1), linetype='dashed') +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14)) +
  labs(x="Abundance cutoff [%]", y="ASV richness proportion left")
cat('\n')

cat("Number of ASVs \n")
sensitivity_abundance %>% 
  dplyr::select(t,grid,host_species,n) %>% 
  spread(host_species,n) %>% 
  filter(t %in% c(0,0.1,0.2,0.5,1)) %>% 
  knitr::kable()
```

```{r}
### filtering out ASVs with less than threshold of relative abundance
rel_reads_threshold <- 0.01
dat_long <- dat %>% 
  gather("asv_ID", "reads", starts_with("ASV")) %>%
  filter(reads >= rel_reads_threshold) 

 dat <- dat_long %>% spread(asv_ID, reads, fill = 0)
```

##

**All ASVs with less than 0.01 relative abundance within a sample were filtered out** \

## Reads distribution {.tabset}

### Distribution
Total number of reads

```{r}
dat %>% 
ggplot(aes(unfiltered_reads)) +
    geom_density(alpha=.4, fill="blue") +
    theme_bw() +
    theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 16)) +
    geom_vline(xintercept = c(5000, 10000, 20000), linetype="dashed") +
    labs(x="Number of Reads",y="Density")

paste("mean:",mean(dat$unfiltered_reads))
paste("median:",median(dat$unfiltered_reads))
paste("range:",range(dat$unfiltered_reads))
```

### Threshold
How many individuals have less than X reads

```{r}
n_total <- nrow(dat)
n_hosts <- NULL
reads_th <- c(seq(0, 50000,1000))
for(th in reads_th) {
  n_calc <- tibble(n = dat %>% 
    filter(unfiltered_reads>th) %>% 
    nrow(),
    n_prop = n/n_total,
                   reads_threshold = th)
  
  n_hosts <- bind_rows(n_hosts, n_calc)
}

# plotting
n_hosts %>% 
ggplot(aes(reads_threshold, n_prop)) +
  geom_line() +
  geom_point() +
  geom_vline(xintercept = c(5000, 10000, 20000), linetype="dashed") +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 16)) +
  labs(x="Number of Reads Cutoff", y="Proportion of hosts left")

# table
n_hosts %>% filter(reads_threshold %in% c(0, 1000, seq(5000,50000,5000))) %>% 
  dplyr::select(reads_threshold, n) %>% 
  knitr::kable()
```

### Threshold by sp.
How many individuals have less than X reads

```{r}
n_total <- dat %>% group_by(host_species) %>% count()
n_hosts <- NULL
reads_th <- c(seq(0, 50000,1000))
for(th in reads_th) {
  n2 <- dat %>% 
    filter(unfiltered_reads>th) %>% 
    group_by(host_species) %>% count()
  n_calc <- tibble(n2,
                   n_prop = n2$n/n_total$n,
                   reads_threshold = th)
  
  n_hosts <- bind_rows(n_hosts, n_calc)
}

# plotting
n_hosts %>% 
ggplot(aes(reads_threshold, n_prop, color=host_species)) +
  geom_line() +
  geom_point() +
  geom_vline(xintercept = c(5000, 10000, 20000), linetype="dashed") +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 16)) +
  labs(x="Number of Reads Cutoff", y="Proportion of hosts left")

# table
n_hosts %>% filter(reads_threshold %in% c(0, 1000, seq(5000,50000,5000))) %>% 
  dplyr::select(reads_threshold, host_species, n) %>%
  spread(host_species, n) %>% 
  knitr::kable()
```

### Threshold by village
How many individuals have less than X reads

```{r}
n_total <- dat %>% group_by(village) %>% count()
n_hosts <- NULL
reads_th <- c(seq(0, 50000,1000))
for(th in reads_th) {
  n2 <- dat %>% 
    filter(unfiltered_reads>th) %>% 
    group_by(village) %>% count()
  n_calc <- tibble(n2,
                   n_prop = n2$n/n_total$n,
                   reads_threshold = th)
  
  n_hosts <- bind_rows(n_hosts, n_calc)
}

# plotting
n_hosts %>% 
ggplot(aes(reads_threshold, n_prop, color=village)) +
  geom_line() +
  geom_point() +
  geom_vline(xintercept = c(5000, 10000, 20000), linetype="dashed") +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 16)) +
  labs(x="Number of Reads Cutoff", y="Proportion of hosts left")

# table
n_hosts %>% filter(reads_threshold %in% c(0, 1000, seq(5000,50000,5000))) %>% 
  dplyr::select(reads_threshold, village, n) %>%
  spread(village, n) %>% 
  knitr::kable()
```

\
```{r}
### filtering out samples with less than threshold of reads
reads_threshold <- 1000
dat %<>% filter(unfiltered_reads >= reads_threshold & !(is.na(unfiltered_reads)))
```

## ASVs Accumulation Curves {.tabset}

**All host individuals with less than 1000 total reads were filtered out** \

```{r, results="asis"}
common_sp <- c("Mus musculus","Rattus rattus","Microgale brevicaudata")
colrs <- c("#f3722c", "#90be6d", "#577590")

for(s in 1:3) {
 cat('###',common_sp[s],'\n','\n')
 dat_accu <- dat %>% filter(host_species==common_sp[s]) %>% 
 mutate(across(starts_with("ASV"),~ .*unfiltered_reads)) %>% 
  dplyr::select(-host_species,-village,-grid,-season,-unfiltered_reads,-sex,-mass) %>% 
  column_to_rownames("host_ID") %>% 
  mutate(across(.cols = everything(),as.integer))
  #select(where(~ any(. != 0)))
#rare_curve <- t(rarefy(dat_accu, sample = seq(1,10001,100), se = FALSE))
rare_curve_2 <- vegan::rarecurve(dat_accu, step=100, tidy = T)
g <- rare_curve_2 %>%  filter(Sample<=20000) %>% 
  ggplot(aes(x=Sample, y=Species, group=Site)) +
  geom_line(color = colrs[s]) +
  geom_vline(xintercept=c(1000,2000))+
  theme_classic() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14),legend.position = "none") +
  labs(x="Number of Reads", y="ASVs Richness")
print(g)
cat('\n','\n')
}
```

## ASVs Prevalence {.tabset}

```{r}
# transforming to long format
dat_long <- dat %>% gather("asv_ID","reads",starts_with("ASV")) %>% 
  filter(reads > 0)
```

### Villages

In how many villages ASVs occur\

```{r}
dat_long %>% group_by(host_species, asv_ID) %>% 
  summarise(prevalence = n_distinct(village)) %>%
  count(host_species,prevalence) %>% 
  left_join(dat_long %>% group_by(host_species) %>% summarise(n_total = n_distinct(asv_ID)), by="host_species") %>% 
  mutate(proportion = n/n_total) %>% 
  mutate(prevalence = as.factor(prevalence)) %>% 
  ggplot(aes(x=host_species, y=proportion, fill=prevalence)) +
  geom_bar(position="fill", stat="identity") +
  theme_classic() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 16), strip.text.x = element_text(size=12)) +
  labs(x="Host Species", y="Proportion")
```

### Seasons

In how many seasons ASVs occur\

```{r}
dat_long %>% group_by(host_species, asv_ID) %>% 
  summarise(prevalence = n_distinct(season)) %>%
  count(host_species,prevalence) %>% 
  left_join(dat_long %>% group_by(host_species) %>% summarise(n_total = n_distinct(asv_ID)), by="host_species") %>% 
  mutate(proportion = n/n_total) %>% 
  mutate(prevalence = as.factor(prevalence)) %>% 
  ggplot(aes(x=host_species, y=proportion, fill=prevalence)) +
  geom_bar(position="fill", stat="identity") +
  theme_classic() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 16), strip.text.x = element_text(size=12)) +
  labs(x="Host Species", y="Proportion")
```

### Land uses

In how many grids ASVs occur\

```{r}
dat_long %>% group_by(host_species, asv_ID) %>% 
  summarise(prevalence = n_distinct(grid)) %>%
  count(host_species,prevalence) %>% 
  left_join(dat_long %>% group_by(host_species) %>% summarise(n_total = n_distinct(asv_ID)), by="host_species") %>% 
  mutate(proportion = n/n_total) %>% 
  mutate(prevalence = as.factor(prevalence)) %>% 
  ggplot(aes(x=host_species, y=proportion, fill=prevalence)) +
  geom_bar(position="fill", stat="identity") +
  theme_classic() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 16), strip.text.x = element_text(size=12)) +
  labs(x="Host Species", y="Proportion")
```

### Land uses by village

In how many grids ASVs occur\

```{r}
dat_long %>% group_by(village,host_species, asv_ID) %>% 
  summarise(prevalence = n_distinct(grid)) %>%
  count(host_species,prevalence) %>% 
  left_join(dat_long %>% group_by(village,host_species) %>% summarise(n_total = n_distinct(asv_ID))) %>% 
  mutate(proportion = n/n_total) %>% 
  mutate(prevalence = as.factor(prevalence)) %>% 
  ggplot(aes(x=host_species, y=proportion, fill=prevalence)) +
  geom_bar(position="fill", stat="identity") +
  facet_wrap(~village) +
  theme_classic() +
  theme(axis.text = element_text(size = 10, color = 'black', angle = 90, vjust = 0.5, hjust=1), title = element_text(size = 16), strip.text.x = element_text(size=12)) +
  labs(x="Host Species", y="Proportion")
```

### Individuals

In how many individuals ASVs occur\
Given by the proportion of individuals out of the total abundance of the focal host species

```{r}
n_hosts <- dat %>% count(host_species, name="n_host")
dat_long %>% group_by(host_species, asv_ID) %>% 
  summarise(prevalence = n_distinct(host_ID)) %>%
  left_join(n_hosts, by="host_species") %>% 
  mutate(prevalence_prop = prevalence/n_host) %>% 
  mutate(prevalence_prop = cut(prevalence_prop, breaks=seq(0,1,0.1))) %>% 
  count(host_species,prevalence_prop) %>% 
  left_join(dat_long %>% group_by(host_species) %>% summarise(n_total = n_distinct(asv_ID)), by="host_species") %>% 
  mutate(proportion = n/n_total) %>% 
  mutate(prevalence_prop = as.factor(prevalence_prop)) %>% 
  ggplot(aes(x=host_species, y=proportion, fill=prevalence_prop)) +
  geom_bar(position="fill", stat="identity") +
  theme_classic() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 16), strip.text.x = element_text(size=12)) +
  labs(x="Host Species", y="Proportion")
```

### Individuals_2

In how many individuals ASVs occur\

```{r}
dat_long %>% group_by(host_species, asv_ID) %>% 
  summarise(prevalence = n_distinct(host_ID)) %>%
  ggplot(aes(x=prevalence)) +
  geom_histogram(binwidth = 1) +
  facet_wrap(~host_species)+
  geom_vline(xintercept = c(10, 50), linetype="dashed") +
  xlim(0,100)+
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14), strip.text.x = element_text(size=12)) +
  labs(x="No. of Host Individuals", y="No. of ASVs")
```


### Individuals by grid

In how many individuals ASVs occur\
Given by the proportion of individuals out of the total abundance of the focal host species in the grid

```{r}
n_hosts <- dat %>% count(grid ,host_species, name="n_host")
dat_long %>% group_by(grid,host_species, asv_ID) %>% 
  summarise(prevalence = n_distinct(host_ID)) %>%
  left_join(n_hosts) %>% 
  mutate(prevalence_prop = prevalence/n_host) %>% 
  mutate(prevalence_prop = cut(prevalence_prop, breaks=seq(0,1,0.1))) %>% 
  count(host_species,prevalence_prop) %>% 
  left_join(dat_long %>% group_by(host_species) %>% summarise(n_total = n_distinct(asv_ID)), by="host_species") %>% 
  mutate(proportion = n/n_total) %>% 
  mutate(prevalence_prop = as.factor(prevalence_prop)) %>% 
  ggplot(aes(x=host_species, y=proportion, fill=prevalence_prop)) +
  geom_bar(position="fill", stat="identity") +
  facet_wrap(~grid) +
  theme_classic() +
  theme(axis.text = element_text(size = 8, color = 'black', angle = 90, vjust = 0.5, hjust=1), title = element_text(size = 16), strip.text.x = element_text(size=12)) +
  labs(x="Host Species", y="Proportion")
```

### Host Species

In how many host species ASVs occur - shared ASVs\

```{r}
dat_long %>% group_by(grid, asv_ID) %>% 
  summarise(prevalence = n_distinct(host_species)) %>%
  count(grid,prevalence) %>% 
  left_join(dat_long %>% group_by(grid) %>% summarise(n_total = n_distinct(asv_ID)), by="grid") %>% 
  mutate(proportion = n/n_total) %>% 
  mutate(prevalence = as.factor(prevalence)) %>% 
  ggplot(aes(x=grid, y=proportion, fill=prevalence)) +
  geom_bar(position="fill", stat="identity") +
  theme_classic() +
  theme(axis.text = element_text(size = 10, color = 'black', angle = 90, vjust = 0.5, hjust=1), title = element_text(size = 16), strip.text.x = element_text(size=12)) +
  labs(x="Land Use", y="Proportion")
```

# Microbes Alpha Diversity

## Alpha Diversity {.tabset}
Calculating average (per host) alpha diversity measures\
**No normalization applied** \

### Observed Richness
Average observed ASVs richness\

```{r}
richness <- dat_long %>% group_by(host_ID) %>% 
  summarise(richness = n_distinct(asv_ID))

#plotting ASVs richness by village, grid, and host species
dat_long %>% 
  group_by(village, grid, host_species, host_ID) %>% 
  summarise(richness = n_distinct(asv_ID)) %>% 
  ggplot(aes(x=grid, y=richness, fill=village)) +
  facet_wrap(~host_species) +
  geom_boxplot() +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black', angle = 90, vjust = 0.5, hjust=1), title = element_text(size = 16), strip.text.x = element_text(size = 12)) +
  labs(x="Land-use", y="Average ASV richness") 
```

### Shannon Diversity
Average Shannon diversity\

```{r}
dat_shannon <- dat %>% 
  dplyr::select(starts_with("ASV")) %>% 
  mutate(across(starts_with("ASV"),~ .*log(.))) 
  shannon <- -rowSums(dat_shannon, na.rm = TRUE)
  
  dat %>% 
    dplyr::select(!starts_with("ASV")) %>% 
    mutate(shannon = shannon) %>% 
  ggplot(aes(x=grid, y=shannon, fill=village)) +
  facet_wrap(~host_species) +
  geom_boxplot() +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black', angle = 90, vjust = 0.5, hjust=1), title = element_text(size = 16), strip.text.x = element_text(size = 12)) +
  labs(x="Land-use", y="Average Shannon Diversity") 

```

### Phylogenetic Diversity
Average phylogenetic species evenness\

```{r}
# reading phylogenetic tree
phylo_tree <- readRDS("output/phylogenetic_tree/phylo_tree_0.01.rds")
  tree <- phylo_tree$tree
  rooted_tree <- midpoint(tree) # rooting the tree
  # **rooted_tree** is the full (after 1% filter) tree I should work with
  
  dat_phylo_matrix <- dat %>% 
  column_to_rownames("host_ID") %>% 
  dplyr::select(starts_with("ASV")) %>% 
  dplyr::select(where(~ any(. > 0, na.rm = TRUE)))
  
  # PSE - take into account phylogenetic distance and species abundance
  # min value: 0; max value: 1
  dat_phylo <- picante::pse(dat_phylo_matrix, rooted_tree)
  dat_phylo %<>% rownames_to_column("host_ID") %>% 
    mutate_at("PSEs", ~replace(., is.na(.), 0)) %>% 
    mutate(host_ID = as.numeric(host_ID)) %>% 
    left_join(dat %>% dplyr::select(host_ID,host_species,village,grid), by="host_ID")
  
  # plotting
  dat_phylo %>% 
    ggplot(aes(x=grid, y=PSEs, fill=village)) +
  facet_wrap(~host_species) +
  geom_boxplot() +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black', angle = 90, vjust = 0.5, hjust=1), title = element_text(size = 16), strip.text.x = element_text(size = 12)) +
  labs(x="Land-use", y="Average phylogenetic Diversity")
```

```{r}
# combining all diversity measures into one data frame + hosts metadata
asv_diversity <- tibble(host_ID = richness$host_ID,
                        richness = richness$richness,
                        shannon = shannon,
                        phylo = dat_phylo$PSEs) %>% 
  left_join(data_sm, by="host_ID")
```

## GLM all variables {.tabset}

Independent variable: ASVs diversity (observed richness, shannon)\
Dependent variables: village, grid, season, sex, mass\
**The best models: delta AICc <= 2**\

```{r}
 for(s in common_sp) {
  cat('###',s,'\n','\n')
  
asv_diversity_host <- asv_diversity %>% filter(host_species == s)
# model formula
if(length(unique(asv_diversity_host$village)) > 1) { 
glm_richness <- glm(richness~village+grid+sex+mass+season, data = asv_diversity_host, family = gaussian, na.action = na.fail)
glm_shannon <- glm(shannon~village+grid+sex+mass+season, data = asv_diversity_host, family = gaussian, na.action = na.fail)
glm_phylo <- glm(phylo~village+grid+sex+mass+season, data = asv_diversity_host, family = gaussian, na.action = na.fail)
}
# microgale occur only in one village
  else {
glm_richness <- glm(richness~grid+sex+mass+season, data = asv_diversity_host, family = gaussian, na.action = na.fail)
glm_shannon <- glm(shannon~grid+sex+mass+season, data = asv_diversity_host, family = gaussian, na.action = na.fail)
glm_phylo <- glm(phylo~grid+sex+mass+season, data = asv_diversity_host, family = gaussian, na.action = na.fail)
}

# AIC for richness
dredge_richness <- MuMIn::dredge(glm_richness)
# The best models: delta <= 2
results_richness <- subset(dredge_richness, delta <= 2 | df == 2 | df == max(df), recalc.weights = FALSE)
row.names(results_richness) <- c(1:(nrow(results_richness)))
cat("ASVs richness -",'\n \n')
print(knitr::kable(results_richness))
# p-value of significant variables in the full model
sig <- as.data.frame(summary(glm_richness)$coef) %>% rownames_to_column() %>% rename('p' = 'Pr(>|t|)') %>% filter(p < 0.05) %>% knitr::kable()
print(sig)
# Relative importance
imp <- as.data.frame(MuMIn::sw(dredge_richness))
var_names <- rownames(imp)
imp_values <- as.vector(imp[[1]])
# Plotting
barplot(imp_values, names.arg=var_names, ylab = "Importance", ylim = c(0,1))
cat('\n')

# AIC for shannon
dredge_shannon <- MuMIn::dredge(glm_shannon)
# The best models: delta <= 2
results_shannon <- subset(dredge_shannon, delta <= 2 | df == 2 | df == max(df), recalc.weights = FALSE)
row.names(results_shannon) <- c(1:(nrow(results_shannon)))
cat("ASVs shannon -",'\n \n')
print(knitr::kable(results_shannon))
# p-value of significant variables in the full model
sig <- as.data.frame(summary(glm_shannon)$coef) %>% rownames_to_column() %>% rename('p' = 'Pr(>|t|)') %>% filter(p < 0.05) %>% knitr::kable()
print(sig)
# Relative importance
imp <- as.data.frame(MuMIn::sw(results_shannon))
var_names <- rownames(imp)
imp_values <- as.vector(imp[[1]])
# Plotting
barplot(imp_values, names.arg=var_names, ylab = "Importance", ylim = c(0,1))
cat('\n')

# AIC for Phylo
dredge_phylo <- MuMIn::dredge(glm_phylo)
# The best models: delta <= 2
results_phylo <- subset(dredge_phylo, delta <= 2 | df == 2 | df == max(df), recalc.weights = FALSE)
row.names(results_phylo) <- c(1:(nrow(results_phylo)))
cat("ASVs Phylo -",'\n \n')
print(knitr::kable(results_phylo))
# p-value of significant variables in the full model
sig <- as.data.frame(summary(glm_phylo)$coef) %>% rownames_to_column() %>% rename('p' = 'Pr(>|t|)') %>% filter(p < 0.05) %>% knitr::kable()
print(sig)
# Relative importance
imp <- as.data.frame(MuMIn::sw(results_phylo))
var_names <- rownames(imp)
imp_values <- as.vector(imp[[1]])
# Plotting
barplot(imp_values, names.arg=var_names, ylab = "Importance", ylim = c(0,1))
cat('\n','\n')

}
```

## GLMM land use {.tabset}

Fit a generalized linear mixed-effects model (GLMM) to check differences in ASVs alpha diversity between land uses\
Fixed variable: grid\
Random variables: village, season\
\
Tukey comparisons:\
1 = semi-intact forest\
2 = secondary_forest\
3 = brushy_regrowth\
4 = agriculture\
5 = agroforest\
6 = flooded_rice\
7 = village\

```{r, include=FALSE}
# Fitting distribution to richness data
asv_div <- asv_diversity %>% filter(host_species == common_sp[2]) %>% dplyr::select(richness)
fit_norm <- fitdistrplus::fitdist(asv_div[[1]], distr = "norm")
fit_nbinom <- fitdist(asv_div[[1]], distr = "nbinom")
fit_lnorm <- fitdist(asv_div[[1]], distr = "lnorm")

par(mfrow = c(2, 2))
plot.legend <- c("Normal", "Neg Binomial","LogNormal")
denscomp(list(fit_norm,fit_nbinom,fit_lnorm), legendtext = plot.legend)
qqcomp(list(fit_norm,fit_nbinom,fit_lnorm), legendtext = plot.legend)
cdfcomp(list(fit_norm,fit_nbinom,fit_lnorm), legendtext = plot.legend)
ppcomp(list(fit_norm,fit_nbinom,fit_lnorm), legendtext = plot.legend)

gofstat(list(fit_norm,fit_nbinom,fit_lnorm),fitnames = c("Normal", "Neg Binomial","LogNormal"))
```

```{r, include=FALSE}
# Fitting distribution to shannon data
asv_div <- asv_diversity %>% filter(host_species == common_sp[3]) %>% dplyr::select(shannon)
fit_norm <- fitdistrplus::fitdist(asv_div[[1]], distr = "norm")
fit_nbinom <- fitdist(asv_div[[1]], distr = "gamma")
fit_lnorm <- fitdist(asv_div[[1]], distr = "lnorm")

par(mfrow = c(2, 2))
plot.legend <- c("Normal", "Poisson","LogNormal")
denscomp(list(fit_norm,fit_nbinom,fit_lnorm), legendtext = plot.legend)
qqcomp(list(fit_norm,fit_nbinom,fit_lnorm), legendtext = plot.legend)
cdfcomp(list(fit_norm,fit_nbinom,fit_lnorm), legendtext = plot.legend)
ppcomp(list(fit_norm,fit_nbinom,fit_lnorm), legendtext = plot.legend)

gofstat(list(fit_norm,fit_nbinom,fit_lnorm),fitnames = c("Normal", "Poisson","LogNormal"))
```

```{r}
# changing grid name for readability
asv_diversity$grid_id = ifelse(asv_diversity$grid=="semi-intact_forest",1,
                               ifelse(asv_diversity$grid=="secondary_forest",2,
                                      ifelse(asv_diversity$grid=="brushy_regrowth",3,
                                             ifelse(asv_diversity$grid=="agriculture",4,
                                                    ifelse(asv_diversity$grid=="agroforest", 5,
                                                           ifelse(asv_diversity$grid=="flooded_rice",6,
                                                                  7))))))
for(s in common_sp) {
  cat('###',s,'\n','\n')
  
asv_diversity_host <- asv_diversity %>% filter(host_species == s) %>% mutate(grid_id = as.factor(grid_id))
# model formula
if(length(unique(asv_diversity_host$village)) > 1) { 
glm_formula <- "grid_id + (1|village) + (1|season)"
} else {
    glm_formula <- "grid_id + (1|season)"
}

# richness
richness_grid <- glmer(paste("richness ~", glm_formula), data = asv_diversity_host, family = gaussian)
tuk_richness <- multcomp::glht(richness_grid, linfct=mcp(grid_id="Tukey"))
cld_richness <- multcomp::cld(tuk_richness)
# plotting
par(mai=c(1,1,1.5,1), no.readonly=TRUE)
plot(cld_richness)
cat('\n')

# Shannon
shannon_grid <- glmer(paste("shannon ~", glm_formula), data = asv_diversity_host, family = gaussian)
tuk_shannon <- multcomp::glht(shannon_grid, linfct=mcp(grid_id="Tukey"))
cld_shannon <- multcomp::cld(tuk_shannon)
# plotting
par(mai=c(1,1,1.5,1), no.readonly=TRUE)
plot(cld_shannon)
cat('\n')

# Phylogenetic diversity
phylo_grid <- glmer(paste("phylo ~", glm_formula), data = asv_diversity_host, family = gaussian)
tuk_phylo <- multcomp::glht(phylo_grid, linfct=mcp(grid_id="Tukey"))
cld_phylo <- multcomp::cld(tuk_phylo)
# plotting
par(mai=c(1,1,1.5,1), no.readonly=TRUE)
plot(cld_phylo)
cat('\n','\n')
}
```

# Beta Diversity

## Community Similarity - Population {.tabset}

Community similarity between land uses in the population level.\
Population = aggregation of all individuals from the same grid.\
In order to mitigate the bias in hosts abundance across grids, I randomly sample min number of individuals (min = no. of individuals in the smallest grid) from all the grids.\
Then, I aggregate all the individuals from the same grid and calculate similarity between grids.\
I repeat this 100 times and average the results.\

```{r, fig.width=12, fig.height=5}
set.seed(99)
for(s in common_sp) {
  cat('###',s,'\n','\n')
  
  # random sampling of individuals by grid
  dat_s <- dat %>%  filter(host_species == s)
  min_grid <- dat_s %>% count(grid) # number of individuals in each grid
  
  # bootstrapping
  dist_pop_jaccard <- matrix(0,nrow(min_grid),nrow(min_grid))
  dist_pop_bray <- matrix(0,nrow(min_grid),nrow(min_grid))
  d_pop_unifrac_uw <- matrix(0,nrow(min_grid),nrow(min_grid))
  d_pop_unifrac_w <- matrix(0,nrow(min_grid),nrow(min_grid))
  for(i in 1:100){
    # randomly sampling hosts 
    dat_random <- dat_s %>% 
      group_by(grid) %>% 
      slice_sample(n = min(min_grid$n))
    
    # agreggating individuals into populations
    # long format
    dat_asv_pop_long <- dat_random %>% 
      gather("asv_ID", "reads", starts_with("ASV")) %>% 
      filter(reads > 0) 
    # ASVs relative abundance in each grid
    asv_pop_reads <- dat_asv_pop_long %>% 
      group_by(grid, asv_ID) %>% 
      summarise(rel_reads = mean(reads))
    # back to wide format
    dat_asv_pop <- dat_asv_pop_long %>% 
      distinct(grid, asv_ID) %>% 
      left_join(asv_pop_reads) %>% 
      spread(asv_ID, rel_reads, fill = 0) %>% 
      column_to_rownames("grid")
    
    # Jaccard
    dat_asv_pop_j <- dat_asv_pop %>% mutate_all(funs(ifelse(.>0,1,0)))
    dist_pop_jaccard_i <- as.matrix(1-vegdist(dat_asv_pop_j, "jaccard"))
    dist_pop_jaccard <- dist_pop_jaccard + dist_pop_jaccard_i
    # Bray-Curtis
    dist_pop_bray_i <- as.matrix(1-vegdist(dat_asv_pop, "bray"))
    dist_pop_bray <- dist_pop_bray + dist_pop_bray_i
    
    ## UniFrac
    # prune the tree
  included_asvs <- unique(dat_asv_pop_long$asv_ID)
  unincluded <- rooted_tree$tip.label[!rooted_tree$tip.label %in% included_asvs]
  pruned <- dendextend::prune(rooted_tree, unincluded)
  unifracs <- GUniFrac(dat_asv_pop, pruned, alpha=c(0, 0.5, 1))$unifracs
  # unweighted
  d_pop_unifrac_uw_i <- 1-unifracs[, , "d_UW"]
  d_pop_unifrac_uw <- d_pop_unifrac_uw + d_pop_unifrac_uw_i
  # weighted
  d_pop_unifrac_w_i <- 1-unifracs[, , "d_1"]
  d_pop_unifrac_w <- d_pop_unifrac_w + d_pop_unifrac_w_i
  }
  
  # plotting jaccard
  dist_pop_jaccard <- dist_pop_jaccard/i
  dist_pop_jaccard[lower.tri(dist_pop_jaccard)] <- NA
  diag(dist_pop_jaccard) <- NA
  dist_pop_jaccard_m <- melt(dist_pop_jaccard)
  g1 <- ggplot(dist_pop_jaccard_m, aes(x = Var1, y = Var2, fill = value, label=round(value,2))) +
    geom_tile() + 
    geom_text()+
    theme_classic()+
    scale_fill_distiller(palette = "Spectral", limits = c(0,1), na.value = "#F2F4F4")+  
    theme(axis.text = element_text(size=12, color='black'), axis.text.x = element_text(angle = 45, hjust = 1)) +
    labs(x="", y="", title = "Jaccard")
  print(g1)
  cat('\n')
  
  # plotting Bray-Curtis
  dist_pop_bray <- dist_pop_bray/i 
  dist_pop_bray[lower.tri(dist_pop_bray)] <- NA
  diag(dist_pop_bray) <- NA
  dist_pop_bray_m <- melt(dist_pop_bray)
  g2 <- ggplot(dist_pop_bray_m, aes(x = Var1, y = Var2, fill = value, label=round(value,2))) +
    geom_tile() + 
    geom_text()+
    theme_classic()+
    scale_fill_distiller(palette = "Spectral", limits = c(0,1), na.value = "#F2F4F4")+
    theme(axis.text = element_text(size=12, color='black'), axis.text.x = element_text(angle = 45, hjust = 1)) +
    labs(x="", y="", title = "Bray-Curtis")
  print(g2)
  cat('\n')
  
  # plotting Un-weighted UniFrac
  d_pop_unifrac_uw <- d_pop_unifrac_uw/i 
  d_pop_unifrac_uw[lower.tri(d_pop_unifrac_uw)] <- NA
  diag(d_pop_unifrac_uw) <- NA
  d_pop_unifrac_uw_m <- melt(d_pop_unifrac_uw)
  g2 <- ggplot(d_pop_unifrac_uw_m, aes(x = Var1, y = Var2, fill = value, label=round(value,2))) +
    geom_tile() + 
    geom_text()+
    theme_classic()+
    scale_fill_distiller(palette = "Spectral", limits = c(0,1), na.value = "#F2F4F4")+
    theme(axis.text = element_text(size=12, color='black'), axis.text.x = element_text(angle = 45, hjust = 1)) +
    labs(x="", y="", title = "Un-weighted UniFrac")
  print(g2)
  cat('\n')
  
  # plotting weighted UniFrac
  d_pop_unifrac_w <- d_pop_unifrac_w/i 
  d_pop_unifrac_w[lower.tri(d_pop_unifrac_w)] <- NA
  diag(d_pop_unifrac_w) <- NA
  d_pop_unifrac_w_m <- melt(d_pop_unifrac_w)
  g2 <- ggplot(d_pop_unifrac_w_m, aes(x = Var1, y = Var2, fill = value, label=round(value,2))) +
    geom_tile() + 
    geom_text()+
    theme_classic()+
    scale_fill_distiller(palette = "Spectral", limits = c(0,1), na.value = "#F2F4F4")+
    theme(axis.text = element_text(size=12, color='black'), axis.text.x = element_text(angle = 45, hjust = 1)) +
    labs(x="", y="", title = "Weighted UniFrac")
  print(g2)
  cat('\n','\n')
  
}
```


## Community Similarity - Distribution {.tabset}

```{r}
for(s in common_sp) {
  cat('###',s,'\n','\n')
  
# preparing the data
dat_beta <- dat %>% filter(host_species == s)
dat_asv_beta <- dat_beta %>% 
  column_to_rownames("host_ID") %>% 
  dplyr::select(starts_with("ASV")) %>% 
  dplyr::select(where(~ any(. > 0, na.rm = TRUE))) 

# Jaccard
dat_asv_j <- dat_asv_beta %>% mutate_all(funs(ifelse(.>0,1,0)))
dist_jaccard <- as.matrix(1-vegdist(dat_asv_j, "jaccard")) 
dist_jaccard[lower.tri(dist_jaccard)] <- NA
diag(dist_jaccard) <- NA
dist_jaccard_m <- melt(dist_jaccard) %>% 
  rename(jaccard = value) %>% 
  filter(!(is.na(jaccard))) %>% 
  left_join(dat_beta %>% dplyr::select(host_ID,grid),by=c("Var1"="host_ID")) %>% 
  left_join(dat_beta %>% dplyr::select(host_ID,grid),by=c("Var2"="host_ID")) %>% 
  filter(grid.x==grid.y)
# Bray-Curtis
dist_bray <- as.matrix(1-vegdist(dat_asv_beta, "bray")) 
dist_bray[lower.tri(dist_bray)] <- NA
diag(dist_bray) <- NA
dist_bray_m <- melt(dist_bray) %>% 
  rename(bray = value) %>% 
  filter(!(is.na(bray))) %>% 
  left_join(dat_beta %>% dplyr::select(host_ID,grid),by=c("Var1"="host_ID")) %>% 
  left_join(dat_beta %>% dplyr::select(host_ID,grid),by=c("Var2"="host_ID")) %>% 
  filter(grid.x==grid.y)

#  UniFrac
 # pruning the tree
included_asvs <- colnames(dat_asv_beta)
  unincluded <- rooted_tree$tip.label[!rooted_tree$tip.label %in% included_asvs]
  pruned <- dendextend::prune(rooted_tree, unincluded)
  unifracs <- GUniFrac(as.matrix(dat_asv_beta), pruned, alpha=c(0, 0.5, 1))$unifracs
  # Weighted UniFrac
  dist_wunif <- 1-unifracs[, , "d_1"]
dist_wunif[lower.tri(dist_wunif)] <- NA
diag(dist_wunif) <- NA
dist_wunif_m <- melt(dist_wunif) %>% 
  rename(wunifrac = value) %>% 
  filter(!(is.na(wunifrac))) %>% 
  left_join(dat_beta %>% dplyr::select(host_ID,grid),by=c("Var1"="host_ID")) %>% 
  left_join(dat_beta %>% dplyr::select(host_ID,grid),by=c("Var2"="host_ID")) %>% 
  filter(grid.x==grid.y)

# UnWeighted UniFrac
  dist_uwunif <- 1-unifracs[, , "d_UW"]
dist_uwunif[lower.tri(dist_uwunif)] <- NA
diag(dist_uwunif) <- NA
dist_uwunif_m <- melt(dist_uwunif) %>% 
  rename(uwunifrac = value) %>% 
  filter(!(is.na(uwunifrac))) %>% 
  left_join(dat_beta %>% dplyr::select(host_ID,grid),by=c("Var1"="host_ID")) %>% 
  left_join(dat_beta %>% dplyr::select(host_ID,grid),by=c("Var2"="host_ID")) %>% 
  filter(grid.x==grid.y)

# combining indeces
dists <- bind_cols(dist_jaccard_m, bray=dist_bray_m$bray) %>% 
  bind_cols(w_unifrac=dist_wunif_m$wunifrac) %>% 
  bind_cols(uw_unifrac=dist_uwunif_m$uwunifrac) %>% 
  gather("index","value",jaccard, bray, w_unifrac, uw_unifrac)
# calculating the mean
dist_mean <- dists %>% 
  group_by(grid.x, index) %>% 
  summarise(mean = mean(value))

# Plotting
g <- dists %>%  
    ggplot(aes(value, fill=index, cut=index)) +
    geom_density(alpha=.4) +
    facet_wrap(~grid.x) +
    theme_bw() +
    theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14), strip.text.x = element_text(size = 12)) +
    xlim(0, 1) +
    geom_vline(data=dist_mean, aes(xintercept=mean, color=index), linetype="dashed")+
    guides(fill=guide_legend(title="Similarity Index")) +
    labs(x="Community similarity",y="Density")
print(g)
cat('\n','\n')

}
```


## PERMANOVA land use {.tabset}

```{r}
set.seed(199)
for(s in common_sp) {
  cat('###',s,'\n','\n')
  
# preparing the data
dat_beta <- dat %>% filter(host_species == s) 

dat_asv_beta <- dat_beta %>% 
  column_to_rownames("host_ID") %>% 
  dplyr::select(starts_with("ASV")) %>% 
  dplyr::select(where(~ any(. > 0, na.rm = TRUE)))
  
# Jaccard
cat("Jaccard \n")
dat_asv_beta_j <- dat_asv_beta %>%  mutate_all(funs(ifelse(.>0,1,0)))
dist_jaccard <- vegdist(dat_asv_beta_j, method = "jaccard") # distance matrix
beta_jaccard <- adonis2(formula = dist_jaccard ~ grid, data = dat_beta) # PERMANOVA
results_a <- beta_jaccard[1,5]

disp_jaccard <- betadisper(dist_jaccard, group=dat_beta$grid) # PERMDISP - dispersion
results_b <- anova(disp_jaccard)[1,5]

results_sum <- tibble(PERMANOVA = results_a, PERMDIST = results_b)
print(knitr::kable(results_sum))
cat('\n')

# PERMANOVA post-hoc
beta_post_jaccard <- RVAideMemoire::pairwise.perm.manova(dist_jaccard, fact = dat_beta$grid, 
    test = "bonferroni", nperm = 999, progress = FALSE)$p.value %>% as.data.frame()
print(knitr::kable(beta_post_jaccard))
cat('\n')
# Tukey test to check pairwise dispersion comparisons
disp_tuk_jaccard <- TukeyHSD(disp_jaccard)$group %>% as.data.frame() %>% dplyr::select("p adj")
print(knitr::kable(disp_tuk_jaccard))
cat('\n')

# checking for outliers (disconnected samples)
dis <- distconnected(dist_jaccard)
if(length(unique(dis)) > 1){
dat_asv_beta <- dat_asv_beta[dis==1,]  # removing them
dat_asv_beta_j <- dat_asv_beta_j[dis==1,]
dat_beta <- dat_beta[dis==1,] 
}

nmds_jaccard <- metaMDS(dat_asv_beta_j, distance = "jaccard", noshare = TRUE, k=2, trace = FALSE) #NMDS
# stressplot(nmds_jaccard)
NMDS1<-nmds_jaccard$points[,1]
NMDS2<-nmds_jaccard$points[,2]
jaccard_plot <- cbind(dat_beta, NMDS1, NMDS2)

# plotting
g <- jaccard_plot %>% 
  ggplot( aes(NMDS1, NMDS2, color=grid)) +
  geom_point(position=position_jitter(.1)) +
  stat_ellipse(aes(fill=grid), alpha=.1, type='norm',linetype =2, geom="polygon") + ##draws 95% confidence interval ellipses
  theme_minimal() +
  annotate("text", x=0, y=max(abs(NMDS2)), label=paste('Stress =',round(nmds_jaccard$stress,3))) # adding stress
print(g)
cat('\n')

# Bray-curtis
cat("Bray-Curtis \n")
dist_bray <- vegdist(dat_asv_beta, method = "bray") # distance matrix
beta_bray <- adonis2(formula = dist_bray ~ grid, data = dat_beta) # PERMANOVA
results_a <- beta_bray[1,5]

disp_bray <- betadisper(dist_bray, group=dat_beta$grid) # PERMDISP - dispersion
results_b <- anova(disp_bray)[1,5]

results_sum <- tibble(PERMANOVA = results_a, PERMDIST = results_b)
print(knitr::kable(results_sum))
cat('\n')

# PERMANOVA post-hoc
beta_post_bray <- RVAideMemoire::pairwise.perm.manova(dist_bray, fact = dat_beta$grid, 
    test = "bonferroni", nperm = 999, progress = FALSE)$p.value %>% as.data.frame()
print(knitr::kable(beta_post_bray))
cat('\n')
# Tukey test to check pairwise dispersion comparisons
disp_tuk_bray <- TukeyHSD(disp_bray)$group %>% as.data.frame() %>% dplyr::select("p adj")
print(knitr::kable(disp_tuk_bray))
cat('\n')

nmds_bray <- metaMDS(dat_asv_beta, distance = "bray", k=2, noshare = TRUE, trace = FALSE) #NMDS
# stressplot(nmds_bray)
NMDS1<-nmds_bray$points[,1]
NMDS2<-nmds_bray$points[,2]
bray_plot <- cbind(dat_beta, NMDS1, NMDS2)

# plotting
g <- bray_plot %>% 
  ggplot( aes(NMDS1, NMDS2, color=grid)) +
  geom_point(position=position_jitter(.1)) +
  stat_ellipse(aes(fill=grid), alpha=.1, type='norm',linetype =2, geom="polygon") + ##draws 95% confidence interval ellipses
  theme_minimal() +
  annotate("text", x=0, y=max(abs(NMDS2)), label=paste('Stress =',round(nmds_bray$stress,3))) # adding stress
print(g)
cat('\n')

# Weighted UniFrac
cat("Weighted UniFrac \n")
# pruning the tree
included_asvs <- colnames(dat_asv_beta)
  unincluded <- rooted_tree$tip.label[!rooted_tree$tip.label %in% included_asvs]
  pruned <- dendextend::prune(rooted_tree, unincluded)
  unifracs <- GUniFrac(as.matrix(dat_asv_beta), pruned, alpha=c(0, 0.5, 1))$unifracs

dist_wunif <- as.dist(1-unifracs[, , "d_1"])
beta_wunif <- adonis2(formula = dist_wunif ~ grid, data = dat_beta) # PERMANOVA
results_a <- beta_wunif[1,5]

disp_wunif <- betadisper(dist_wunif, group=dat_beta$grid) # PERMDISP - dispersion
results_b <- anova(disp_wunif)[1,5]

results_sum <- tibble(PERMANOVA = results_a, PERMDIST = results_b)
print(knitr::kable(results_sum))
cat('\n')

# PERMANOVA post-hoc
beta_post_wunif <- RVAideMemoire::pairwise.perm.manova(dist_wunif, fact = dat_beta$grid, 
    test = "bonferroni", nperm = 999, progress = FALSE)$p.value %>% as.data.frame()
print(knitr::kable(beta_post_wunif))
cat('\n')
# Tukey test to check pairwise dispersion comparisons
disp_tuk_wunif <- TukeyHSD(disp_wunif)$group %>% as.data.frame() %>% dplyr::select("p adj")
print(knitr::kable(disp_tuk_wunif))
cat('\n')

nmds_wunif <- metaMDS(dist_wunif, k=2, trace = FALSE) #NMDS
# stressplot(nmds_wunif)
NMDS1<-nmds_wunif$points[,1]
NMDS2<-nmds_wunif$points[,2]
wunif_plot <- cbind(dat_beta, NMDS1, NMDS2)

# plotting
g <- wunif_plot %>% 
  ggplot( aes(NMDS1, NMDS2, color=grid)) +
  geom_point(position=position_jitter(.1)) +
  stat_ellipse(aes(fill=grid), alpha=.1, type='norm',linetype =2, geom="polygon") + ##draws 95% confidence interval ellipses
  theme_minimal() +
  annotate("text", x=0, y=max(abs(NMDS2)), label=paste('Stress =',round(nmds_wunif$stress,3))) # adding stress
print(g)
cat('\n','\n')
}
```


