---
title: "Deterministic and stochastic processes drive host-microbe network structures along a land use change gradient"
subtitle: "Analysis and results - core/non-core"
author: "Matan Markfeld"
date: "Last edit: 2024-12-23"
output: 
  html_document:
    toc: true
    toc_float: true
    collapse: false
    number_sections: true
    code_folding: hide
    highlight: tango
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(results = "asis", message=FALSE, warning=FALSE, cache=TRUE, eval = TRUE, dev = c('png','pdf'), out.width = '100%', out.height='100%')
```


```{r load libraries, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(magrittr)
library(ggplot2)
library(vegan)
library(igraph)
library(reshape2)
library(infomapecology)
library(aricode)
library(ape)
library(picante)
library(patchwork)
library(ggrepel)


rm(list=ls())
```

```{r}
source("../modularity_analysis_functions.R", local = knitr::knit_global())
```

```{r, microbiome data}
# reading the microbiome data
# working only with Rattus from the three villages

data_asv <- read_csv("../../data/data_processed/microbiome/data_asv_rra0.001_p0.01_th5000_all.csv")
```

```{r, main analysis, include=FALSE, results='hide'}
# first of all, run this chunk of code.
# this code does all the modularity analysis using the functions in "modularity_analysis_functions".
# the output of the code is variables holding the final results tables and figures (the output of the functions used)

# thresholds for core
core_th <- seq(0.05, 0.5, by=0.05)

group.colors <- c(Core = "#d95f02", 'Non-core' = "#7570b3", Rare = "#2b9348", All = "black")


modules_similarity2_three_groups <- NULL
modules_table_three_groups <- NULL
assembly_three_groups <- NULL

# reading the phylogenetic tree
best_tree <- readRDS(file = "../../results/phylo_tree_rra0.001_p0.01.rds")
phylo_tree <- best_tree$tree 
# ASVs phylogenetic distance
asv_distance <- ape::cophenetic.phylo(phylo_tree)

for (th in core_th) {
  
data_asv_p <- data_asv %>% 
  mutate(asv_core = case_when(host_p>th ~ "Core",
                                          .default = "Non-core"))
core_names <- unique(data_asv_p$asv_core)


# loop for three groups
for (v in core_names) {
  
  
  
  data_asv_group <- data_asv_p %>% 
    filter(asv_core == v)
  
  # calculating ASVs degree
  asv_degree <- data_asv_group %>%
    group_by(asv_ID) %>%
    summarise(n = n_distinct(host_ID)) %>%
    dplyr::rename(asv_degree = n)
  data_asv_group %<>% left_join(asv_degree, by="asv_ID")
  
  ##### observed network
  # finding modules
  modules_observed <- fun_modularity_analysis(data_asv_group) %>% 
    mutate(threshold=th)
  modules_table_three_groups <- rbind(modules_table_three_groups, modules_observed)
  #write_csv(modules_table_three_groups, "../results/modules_table_sarahandrano.csv")
  
  
  # calculating similarity in modules
  modules_similarity2 <- fun_modules_similarity2(modules_observed) 
  # %>% filter((host_group!=1 & asv_core!="Rare") | asv_core=="Rare")
  # saving results in one table
  modules_similarity2_three_groups <- append(modules_similarity2_three_groups, modules_similarity2)
  
  
  ##### phylogenetic analysis
  # ASVs pool
  asv_pool <- data_asv_group %>% 
    distinct(asv_ID, asv_degree) %>% 
    mutate(p = asv_degree/length(unique(data_asv_group$host_ID)))
  
  data_assembly <- data_asv_group %>% 
    mutate(reads=1) %>% 
    distinct(host_ID, asv_ID, reads) %>% 
    spread(asv_ID, reads, fill = 0) %>% 
    column_to_rownames("host_ID") %>% 
    as.matrix()
  
  
  asv_distance2 <- asv_distance[rownames(asv_distance)%in%asv_pool$asv_ID , colnames(asv_distance)%in%asv_pool$asv_ID]
  
  raupc <- as.matrix(raupcrick(data_assembly, nsimul = 100))
  assembly <- iCAMP::qpen(data_assembly, pd = asv_distance2, RC=raupc, rand=100, sig.bNTI = 2, sig.rc = 0.95, ab.weight=F, nworker = 4)
  
  assembly_result <- assembly$result %>% 
    mutate(sample1 = as.double(sample1), sample2 = as.double(sample2)) %>% 
    left_join(modules_observed %>% 
                distinct(host_ID, host_group, grid), by=c("sample1"="host_ID")) %>% 
    dplyr::rename(host_group1=host_group, grid1=grid) %>% 
    left_join(modules_observed %>% 
                distinct(host_ID, host_group, grid), by=c("sample2"="host_ID")) %>% 
    dplyr::rename(host_group2=host_group, grid2=grid) %>% 
    mutate(threshold = th, asv_core = v)
  
  assembly_three_groups <- rbind(assembly_three_groups, assembly_result)

  
}  
}
```

# Network exploration

```{r, network_exploration}
# number of ASVs
group_n <- modules_table_three_groups %>% 
  group_by(threshold, asv_core) %>% 
  summarise(asv_n = n_distinct(asv_ID), host_n = n_distinct(host_ID))

# connectance
interaction_n <- modules_table_three_groups %>% 
  count(threshold, asv_core)
group_n %<>% left_join(interaction_n, by=c("threshold","asv_core")) %>% 
  mutate(connectance = n / (asv_n*host_n))

# number of modules
modules_n <- modules_table_three_groups %>% 
  group_by(threshold, asv_core) %>% 
  summarise(modules_n = n_distinct(host_group))

# mean modules size
modules_size <- modules_table_three_groups %>% 
  distinct(threshold, asv_core, host_group, host_ID) %>% 
  count(threshold, asv_core, host_group) %>% 
  group_by(threshold, asv_core) %>% 
  summarise(modules_size = mean(n), largest_module = max(n))
modules_n %<>% left_join(modules_size, by=c("threshold","asv_core")) 

final_network_properties <- group_n %>% 
  left_join(modules_n, by=c("threshold","asv_core"))

# plotting
g <- final_network_properties %>% 
  select(-n, -host_n) %>% 
  pivot_longer(cols=asv_n:largest_module, names_to="measure", values_to="value") %>% 
  mutate(measure = case_when(measure=="asv_n" ~ "No. of ASVs",
                                               measure=="connectance" ~ "Connectance",
                                               measure=="modules_n" ~ "No. of Modules",
                                               measure=="modules_size" ~ "Avg. Module Size",
                             measure=="largest_module" ~ "Largest Module")) %>% 
    mutate(measure = factor(measure, levels = c("No. of ASVs","Connectance","No. of Modules","Avg. Module Size","Largest Module"))) %>%
  ggplot(aes(x=threshold, y=value, color=asv_core)) +
  geom_line() +
  geom_vline(xintercept = 0.2, linetype = "dotted", color = "grey")+
  facet_wrap(~measure, scales="free", nrow = 1) +
  theme_bw() +
  theme(axis.text = element_text(size = 9, color = 'black', angle = 90, vjust = 0.5, hjust=1), title = element_text(size = 14), 
        legend.position = "none",
        strip.text = element_text(size=10, color = 'black'), strip.background = element_rect(color = "grey80", size = 1), 
        panel.grid = element_blank(), panel.background = element_rect(colour = "black"), aspect.ratio = 1.7) +
  scale_color_manual(values=group.colors) +
  labs(x="Core Threshold", y="Value", color="Microbial Group")
print(g)

```



# assembly processes analysis

```{r, assembly_processes}
assembly_final <- assembly_three_groups %>% 
  mutate(same_module = ifelse(host_group1==host_group2, "Same Module","Different Module"))


# renaming the process
assembly_final %<>% mutate(process = case_when(process=="Heterogeneous.Selection" ~ "Heterogeneous Selection",
                                               process=="Homogeneous.Selection" ~ "Homogeneous Selection",
                                               process=="Dispersal.Limitation" ~ "Dispersal Limitation",
                                               process=="Homogenizing.Dispersal" ~ "Homogenizing Dispersal",
                                               .default = "Drift"))

# summary
assembly_summary <- assembly_final %>% 
  count(threshold, asv_core, same_module, process)

assembly_summary_total_module <- assembly_summary %>% 
  group_by(threshold, asv_core, same_module) %>% 
  summarise(n_total = sum(n))


#plotting process ratio between modules
g <- assembly_summary %>% 
  left_join(assembly_summary_total_module, by=c("threshold","asv_core","same_module")) %>% 
  mutate(n_p = n/n_total) %>% 
  ggplot(aes(fill=process, x=threshold, y=n_p)) +
  geom_area() +
  facet_grid(cols = vars(asv_core), rows = vars(same_module)) +
  #geom_vline(xintercept = 0.3, linetype = "dotted", color = "grey")+
  geom_hline(yintercept = 0.5, linetype = "dotted", color = "black")+
  scale_y_continuous(labels = scales::percent) +
  theme_bw() +
  theme(axis.text = element_text(size = 9, color = 'black', angle = 90, vjust = 0.5, hjust=1), title = element_text(size = 14),
        strip.text = element_text(size=12, color = 'black'), strip.background = element_rect(color = "grey80", size = 1), 
        panel.grid = element_blank(), panel.background = element_rect(colour = "black")) +
  scale_fill_manual(values=c("#a7c957","#f2e8cf","#f07167","#83c5be")) +
  labs(x="Core Threshold", y="Percentage", fill="Process")
print(g)


g <- assembly_summary %>% 
  left_join(assembly_summary_total_module, by=c("threshold","asv_core","same_module")) %>% 
  mutate(n_p = n/n_total) %>% 
  ggplot(aes(fill=process, x=threshold, y=n_p)) +
  geom_bar(position="fill", stat="identity") +
  facet_grid(cols = vars(same_module), rows = vars(asv_core)) +
  scale_y_continuous(labels = scales::percent) +
  geom_text(aes(label = paste0(round(n_p*100,1),"%")), 
            position = position_stack(vjust = 0.5), size = 2.5)+
  theme_bw() +
  theme(axis.text = element_text(size = 9, color = 'black', angle = 90, vjust = 0.5, hjust=1), title = element_text(size = 14),
        legend.position = "bottom",
        strip.text = element_text(size=12, color = 'black'), strip.background = element_rect(color = "grey80", size = 1), 
        panel.grid = element_blank(), panel.background = element_rect(colour = "black")) +
  scale_fill_manual(values=c("#a7c957","#f2e8cf","#f07167","#83c5be")) +
  labs(x="Core Threshold", y="Percentage", fill="Process")
print(g)

```



# Modules similarity across land use change gradient

## Correlations between variables

Distance is a proxy for habitat degradation.

```{r, variables_correlation}
# reading grid similarity results
grids_similarity_attr <- read_csv("../../data/data_processed/village_summary_new.csv")
rownames(grids_similarity_attr) <- grids_similarity_attr$grid_village

# correlations between variables
print(psych::pairs.panels(grids_similarity_attr %>%  select(-grid_village,-village,-grid), ellipses = F, lm = F))
cat('\n','\n')
```

```{r}
# RDA for modules

par_names <- expand_grid(core_th, core_names)
anova_model_all <- NULL
anova_rda_all <- NULL
set.seed(123)

# loop for three asv groups
for (i in 1:nrow(par_names)) {
  
  modules_similarity2 <- modules_similarity2_three_groups[[i]]
  # filtering the largest module
  largest_module <- which.max(colSums(modules_similarity2))
  modules_similarity2 <- modules_similarity2[,-largest_module]
  
  grid_names <- rownames(modules_similarity2)
  # filtering matrix to the existing grids
  grids_similarity_attr2 <- grids_similarity_attr %>%  filter(grid_village %in% grid_names & !grepl("village", grid_village))
  modules_similarity2 <- modules_similarity2[rownames(grids_similarity_attr2),]
  
  
  
  n=data_asv %>% group_by(village,grid) %>% summarise(n=n_distinct(host_ID))
  grids_similarity_attr2 %<>% left_join(n, by=c("village","grid"))
  # standardizing variables to zero mean and unit variance
  grids_similarity_attr2[4:11] <- decostand(grids_similarity_attr2[4:11], 'standardize')
  
  # tb-RDA
  
  # hellinger transformation
  modules_similarity2_hell <- decostand(modules_similarity2, 'hell')
  
  
  rda_result <- rda(modules_similarity2_hell ~ veg_PC1+veg_PC2+dist_to_village+elevation+Condition(n), data=grids_similarity_attr2, na.action = na.omit)
  # **Condition(n) - control for the number of host in each village-grid.
  
  # variation explained
  constrained_eig <- t(as.data.frame(rda_result$CCA$eig/rda_result$tot.chi*100))  # for RDAs
  
R2.obs <- RsquareAdj (rda_result)$adj.r.squared

# significance test for the whole model
  anova_model <- anova(rda_result)[1,] %>% 
    mutate(R2.adj = R2.obs, `Pr(>F)` = as.double(`Pr(>F)`), asv_core = par_names[[i,2]], core_th = par_names[[i,1]]) %>% 
    mutate(sig = ifelse(`Pr(>F)` <= 0.05, 1, 0))
  
  anova_model_all <- rbind(anova_model_all, anova_model)
  
  # significance test for all variables
  anova_rda <- anova(rda_result, by = "margin", permutations = 999) %>% 
    mutate(asv_core = par_names[[i,2]], core_th = par_names[[i,1]]) 
  anova_rda$`Pr(>F)` <- p.adjust (anova_rda$`Pr(>F)`, method = 'holm') 
    anova_rda %<>% mutate(`Pr(>F)` = as.double(`Pr(>F)`), sig = ifelse(`Pr(>F)` <= 0.05, 1, 0)) 
  
  anova_rda_all <- rbind(anova_rda_all, anova_rda)
  
}

# plotting

# arranging the tables
anova_model_all_plot <- anova_model_all %>% 
   mutate(R2.adj = ifelse(R2.adj<0,0,R2.adj))

anova_rda_all_plot <- anova_rda_all %>% 
rownames_to_column("var") %>% 
  left_join(anova_model_all_plot %>% filter(sig==1) %>% select(core_th,asv_core,R2.adj), by=c("asv_core","core_th")) %>% 
  filter(sig == 1 & !is.na(R2.adj)) %>% 
  # mutate(var = case_when(grep("veg_PC1",var) ~ "Vegetation1",
  #                                              grep("veg_PC2",var) ~ "Vegetation2",
  #                                              grep("elevation",var) ~ "Elevation",
  #                                              grep("dist_to_village",var) ~ "Distance"))
  group_by(core_th, asv_core, R2.adj) %>%
  summarise(var_combined = paste(var, collapse = "\n"), .groups = "drop")

 g <- anova_model_all_plot %>% 
   ggplot(aes(y=R2.adj, x=core_th, color = asv_core)) +
   geom_vline(xintercept = 0.2, linetype = "dotted", color = "grey")+
  geom_point(aes(shape = as.factor(sig), size = 3)) +
  geom_line()+
  scale_size(guide="none") +
  scale_shape_manual(values = c(1, 16), guide="none") +
  scale_y_continuous(limits = c(0, 0.2)) +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14), 
         #legend.position = "bottom",
        panel.grid = element_blank(), panel.border = element_rect(color = "black"), aspect.ratio = 0.5) +
  scale_color_manual(values=group.colors) +
  labs(x = "Core Threshold", y = bquote("tb-RDA " * R[Adjusted]^2 * ""), color = "Microbial Group")
 print(g)
cat('\n','\n')

# plotting tiles

# geom_label_repel(
#     data = anova_rda_all_plot, 
#     inherit.aes = FALSE, 
#     aes(x = core_th, y = R2.adj, label = var_combined), 
#     seed = 123,
#     size = 3, 
#     direction = "y",
#     box.padding = 0.4,
#     point.padding = 0.4,
#     nudge_y = 0.02,
#     nudge_x = 0.01,
#     min.segment.length = 0.01,
#     arrow = arrow(length = unit(0.01, "npc")),
#     segment.colour = "gray") +
# library(ggnewscale)
# 
# g <- anova_model_all %>% 
#    mutate(R2.adj = ifelse(R2.adj<0,0,R2.adj)) 
#   ggplot() +
#     # Core
#   geom_tile(data = filter(g, asv_core == "Core"), 
#             aes(x = asv_core, y = core_th, fill = R2.adj)) +
#   scale_fill_gradient(low = "white", high = group.colors[1], name = "R2 adj. - Core", limits = c(0, 0.2)) +
#     
#   # Non-core
#   new_scale_fill() +
#   geom_tile(data = filter(g, asv_core == "Non-core"), 
#             aes(x = asv_core, y = core_th, fill = R2.adj)) +
#   scale_fill_gradient(low = "white", high = group.colors[2], name = "R2 adj. - Non-core", limits = c(0, 0.2)) +
# geom_text(
#     data = anova_rda_all_plot,
#     aes(x = asv_core, y = core_th, label = var),
#     size = 3, color = "black")+  
#     theme_bw() +
#   labs(x = "", y = "Core Threshold") +
#   theme(legend.position = "right")
#   
#   
#    ggplot(data = aes(y=core_th, x=asv_core, fill = R2.adj)) +
#   geom_tile(color='white') +
#     theme_classic() +
#     scale_fill_viridis_c(limits = c(0, 0.2)) +
#   theme_bw() +
#   theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14), 
#          legend.position = "bottom",
#         panel.grid = element_blank(), panel.border = element_rect(color = "black")) +
#   scale_color_manual(values=group.colors) +
#   labs(x = "Core Threshold", y = "R2 adj.", color = "Microbial Group")
#  print(g)
# cat('\n','\n')


print(knitr::kable(anova_model_all))
cat('\n','\n')

print(knitr::kable(anova_rda_all))
cat('\n','\n')

```
