---
title: "Deterministic and stochastic processes drive host-microbe network structures along a land use change gradient"
subtitle: "Analysis and results - core/non-core"
author: "Matan Markfeld"
date: "Last edit: 2024-12-23"
output: 
  html_document:
    toc: true
    toc_float: true
    collapse: false
    number_sections: true
    code_folding: hide
    highlight: tango
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(results = "asis", message=FALSE, warning=FALSE, cache=TRUE, eval = TRUE, dev = c('png'), out.width = '100%', out.height='60%')
```


```{r load libraries, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(magrittr)
library(ggplot2)
library(vegan)
library(igraph)
library(reshape2)
library(infomapecology)
library(aricode)
library(ape)
library(picante)
library(patchwork)

rm(list=ls())
```

```{r}
source("../modularity_analysis_functions.R", local = knitr::knit_global())
```

```{r, microbiome data}
# reading the microbiome data
# working only with Rattus from the three villages

data_asv <- read_csv("../../data/data_processed/microbiome/data_asv_rra0.001_p0.02_th5000_all.csv")
```

```{r, main analysis, include=FALSE, results='hide'}
# first of all, run this chunk of code.
# this code does all the modularity analysis using the functions in "modularity_analysis_functions".
# the output of the code is variables holding the final results tables and figures (the output of the functions used)

# thresholds for core
core_th <- seq(0.05, 0.5, by=0.05)

group.colors <- c(Core = "#f4a261", 'Non-core' = "#0077b6", Rare = "#2b9348", All = "black")


modules_similarity2_three_groups <- NULL
modules_table_three_groups <- NULL
assembly_three_groups <- NULL

# reading the phylogenetic tree
best_tree <- readRDS(file = "../../results/phylo_tree_rra0.001_p0.01.rds")
phylo_tree <- best_tree$tree 
# ASVs phylogenetic distance
asv_distance <- ape::cophenetic.phylo(phylo_tree)

for (th in core_th) {
  
data_asv_p <- data_asv %>% 
  mutate(asv_core = case_when(host_p>th ~ "Core",
                                          .default = "Non-core"))
core_names <- unique(data_asv_p$asv_core)


# loop for three groups
for (v in core_names) {
  
  
  
  data_asv_group <- data_asv_p %>% 
    filter(asv_core == v)
  
  # calculating ASVs degree
  asv_degree <- data_asv_group %>%
    group_by(asv_ID) %>%
    summarise(n = n_distinct(host_ID)) %>%
    dplyr::rename(asv_degree = n)
  data_asv_group %<>% left_join(asv_degree, by="asv_ID")
  
  ##### observed network
  # finding modules
  modules_observed <- fun_modularity_analysis(data_asv_group) %>% 
    mutate(threshold=th)
  modules_table_three_groups <- rbind(modules_table_three_groups, modules_observed)
  #write_csv(modules_table_three_groups, "../results/modules_table_sarahandrano.csv")
  
  
  # calculating similarity in modules
  modules_similarity2 <- fun_modules_similarity2(modules_observed %>% filter((host_group!=1 & asv_core!="Rare") | asv_core=="Rare"))
  # saving results in one table
  modules_similarity2_three_groups <- append(modules_similarity2_three_groups, modules_similarity2)
  
  
  ##### phylogenetic analysis
  # ASVs pool
  asv_pool <- data_asv_group %>% 
    distinct(asv_ID, asv_degree) %>% 
    mutate(p = asv_degree/length(unique(data_asv_group$host_ID)))
  
  data_assembly <- data_asv_group %>% 
    mutate(reads=1) %>% 
    distinct(host_ID, asv_ID, reads) %>% 
    spread(asv_ID, reads, fill = 0) %>% 
    column_to_rownames("host_ID") %>% 
    as.matrix()
  
  
  asv_distance2 <- asv_distance[rownames(asv_distance)%in%asv_pool$asv_ID , colnames(asv_distance)%in%asv_pool$asv_ID]
  
  raupc <- as.matrix(raupcrick(data_assembly, nsimul = 50))
  assembly <- iCAMP::qpen(data_assembly, pd = asv_distance2, RC=raupc, rand=50, sig.bNTI = 2, sig.rc = 0.95, ab.weight=F, nworker = 4)
  
  assembly_result <- assembly$result %>% 
    mutate(sample1 = as.double(sample1), sample2 = as.double(sample2)) %>% 
    left_join(modules_observed %>% distinct(host_ID, host_group, grid), by=c("sample1"="host_ID")) %>% dplyr::rename(host_group1=host_group, grid1=grid) %>% 
    left_join(modules_observed %>% distinct(host_ID, host_group, grid), by=c("sample2"="host_ID")) %>% dplyr::rename(host_group2=host_group, grid2=grid) %>% 
    mutate(threshold = th, asv_core = v)
  
  assembly_three_groups <- rbind(assembly_three_groups, assembly_result)

  
}  
}
```

# Network exploration

```{r, network_exploration}
# number of ASVs
group_n <- modules_table_three_groups %>% 
  group_by(threshold, asv_core) %>% 
  summarise(asv_n = n_distinct(asv_ID), host_n = n_distinct(host_ID))

# connectance
interaction_n <- modules_table_three_groups %>% 
  count(threshold, asv_core)
group_n %<>% left_join(interaction_n, by=c("threshold","asv_core")) %>% 
  mutate(connectance = n / (asv_n*host_n))

# number of modules
modules_n <- modules_table_three_groups %>% 
  group_by(threshold, asv_core) %>% 
  summarise(modules_n = n_distinct(host_group))

# mean modules size
modules_size <- modules_table_three_groups %>% 
  distinct(threshold, asv_core, host_group, host_ID) %>% 
  count(threshold, asv_core, host_group) %>% 
  group_by(threshold, asv_core) %>% 
  summarise(modules_size = mean(n), largest_module = max(n))
modules_n %<>% left_join(modules_size, by=c("threshold","asv_core")) 

final_network_properties <- group_n %>% 
  left_join(modules_n, by=c("threshold","asv_core"))

# plotting
g <- final_network_properties %>% 
  select(-n, -host_n) %>% 
  pivot_longer(cols=asv_n:largest_module, names_to="measure", values_to="value") %>% 
  mutate(measure = case_when(measure=="asv_n" ~ "No. of ASVs",
                                               measure=="connectance" ~ "Connectance",
                                               measure=="modules_n" ~ "No. of Modules",
                                               measure=="modules_size" ~ "Avg. Module Size",
                             measure=="largest_module" ~ "Largest Module")) %>% 
    mutate(measure = factor(measure, levels = c("No. of ASVs","Connectance","No. of Modules","Avg. Module Size","Largest Module"))) %>%
  ggplot(aes(x=threshold, y=value, color=asv_core)) +
  geom_line() +
  geom_vline(xintercept = 0.3, linetype = "dotted", color = "grey")+
  facet_wrap(~measure, scales="free", nrow = 1) +
  theme_bw() +
  theme(axis.text = element_text(size = 9, color = 'black', angle = 90, vjust = 0.5, hjust=1), title = element_text(size = 14), 
        legend.position = "bottom",
        strip.text = element_text(size=10, color = 'black'), strip.background = element_rect(color = "grey80", size = 1), 
        panel.grid = element_blank(), panel.background = element_rect(colour = "black")) +
  scale_color_manual(values=group.colors) +
  labs(x="Core Threshold", y="Value", color="Microbial Group")
print(g)

```



# assembly processes analysis

```{r, assembly_processes}
assembly_final <- assembly_three_groups %>% 
  mutate(same_module = ifelse(host_group1==host_group2, "Same Module","Different Module"))


# renaming the process
assembly_final %<>% mutate(process = case_when(process=="Heterogeneous.Selection" ~ "Heterogeneous Selection",
                                               process=="Homogeneous.Selection" ~ "Homogeneous Selection",
                                               process=="Dispersal.Limitation" ~ "Dispersal Limitation",
                                               process=="Homogenizing.Dispersal" ~ "Homogenizing Dispersal",
                                               .default = "Drift"))

# summary
assembly_summary <- assembly_final %>% 
  count(threshold, asv_core, same_module, process)

assembly_summary_total_module <- assembly_summary %>% 
  group_by(threshold, asv_core, same_module) %>% 
  summarise(n_total = sum(n))


#plotting process ratio between modules
g <- assembly_summary %>% 
  left_join(assembly_summary_total_module, by=c("threshold","asv_core","same_module")) %>% 
  mutate(n_p = n/n_total) %>% 
  ggplot(aes(fill=process, x=threshold, y=n_p)) +
  geom_bar(position="fill", stat="identity") +
  facet_grid(cols = vars(same_module), rows = vars(asv_core)) +
  scale_y_continuous(labels = scales::percent) +
  geom_text(aes(label = paste0(round(n_p*100,1),"%")), 
            position = position_stack(vjust = 0.5), size = 2.5)+
  theme_bw() +
  theme(axis.text = element_text(size = 9, color = 'black', angle = 90, vjust = 0.5, hjust=1), title = element_text(size = 14),
        legend.position = "bottom",
        strip.text = element_text(size=12, color = 'black'), strip.background = element_rect(color = "grey80", size = 1), 
        panel.grid = element_blank(), panel.background = element_rect(colour = "black")) +
  scale_fill_manual(values=c("#a7c957","#f2e8cf","#f07167","#83c5be")) +
  labs(x="Core Threshold", y="Percentage", fill="Process")
print(g)

```



# Modules similarity across land use change gradient

## Correlations between variables

Distance is a proxy for habitat degradation.

```{r, variables_correlation}
# reading grid similarity results
grids_similarity_attr <- read_csv("../data/data_processed/village_summary.csv")
```

## Mantel test

```{r, mantel_test}
# mantel test for: modules similarity ~ distance between grids

# distance between grids matrices
load("../data/data_processed/distance_three_villages.RData")
names(grid_distance_three_villages) <- c("Mandena", "Sarahandrano", "Andatsakala")
grid_distance <- grid_distance_three_villages[[vil]]

mantel_result_three_groups <- NULL
final_data_three_groups <- NULL

for (i in seq(0,(length(ths)*3-1),by=3)) {
  
# loop for three asv groups
for (j in 1:3) {
  
  modules_similarity <- modules_similarity_three_groups[[i+j]]
  grid_names <- rownames(modules_similarity)
  # filtering the distance matrix to the existing grids
  grid_distance <- grid_distance[grid_names, grid_names]
  
  # calculating mantel test
  mantel_result <- vegan::mantel(modules_similarity, grid_distance)
  mantel_result_df <- data_frame(r = round(mantel_result$statistic,2),
                                 pvalue = round(1-mantel_result$signif,4),
                                 asv_core = core_names[j],
                                 i = ths[ceiling((i+j)/3)])
  mantel_result_three_groups <- rbind(mantel_result_three_groups, mantel_result_df)
  
}  
}

# plotting r
g1 <- mantel_result_three_groups %>% 
  mutate(sig = ifelse(pvalue<=0.05,1,0)) %>% 
  ggplot(aes(y=r, x=i, color = asv_core)) +
  geom_point(aes(shape = as.factor(sig), size = 3)) +
  geom_line()+
  scale_size(guide="none") +
  scale_shape_manual(values = c(1, 16), guide="none") +
  #scale_y_continuous(limits = c(0.0, 0.9)) +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), title = element_text(size = 14), 
        panel.grid = element_blank(), panel.border = element_rect(color = "black")) +
  scale_color_manual(values=group.colors) +
  labs(x = "Cutoff Change [+]", y = "r (Mantel)", color = "ASV Group")
print(g1)
cat('\n','\n')
```



