---
title: "Differential assembly of core and non-core host-microbe network structures along a land-use change gradient"
subtitle: "Analysis and results - continuous core/non-core threshold"
author: "Matan Markfeld"
date: "Last edit: 2025-04-03"
output: 
  html_document:
    toc: true
    toc_float: true
    collapse: false
    number_sections: true
    code_folding: hide
    highlight: tango
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(results = "asis", message=FALSE, warning=FALSE, cache=TRUE, eval = TRUE, dev = c('png','pdf'), out.width = '100%', out.height='100%')
```


```{r load libraries, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(magrittr)
library(ggplot2)
library(vegan)
library(igraph)
library(reshape2)
library(infomapecology)
library(ape)
library(iCAMP)
library(patchwork)
library(RColorBrewer)

rm(list=ls())
```

```{r}
source("functions.R", local = knitr::knit_global())
```

```{r, microbiome data}
# reading the microbiome data
# working only with Rattus from the three villages

data_asv <- read_csv("../data/data_processed/data_asv_rra0.001_p0.01_th5000_all.csv")

# reading the phylogenetic tree
best_tree <- readRDS(file = "../data/data_processed/phylo_tree_rra0.001_p0.01.rds")
phylo_tree <- best_tree$tree 
# ASVs phylogenetic distance
asv_distance <- ape::cophenetic.phylo(phylo_tree)
```

```{r, main analysis, include=FALSE, results='hide'}
# first of all, run this chunk of code.
# this code does all the analysis using the functions in "functions".
# the output of the code is variables holding the final results tables and figures (the output of the functions used)

# ------------------------------------------------------------------------------
# thresholds for core
core_th <- seq(0.05, 0.5, by=0.05)
# number of simulations to the assembly analysis
n_sim <- 2

# group colors
group.colors <- c(Core = "#d95f02", 'Non-core' = "#034e7b")

# final tables 
modules_similarity_two_groups <- NULL
modules_table_two_groups <- NULL
assembly_two_groups <- NULL
# ------------------------------------------------------------------------------

# loop for core thresholds
for (th in core_th) {
  
  # defining core/non-core ASVs based on the threshold (th)  
  data_asv_p <- data_asv %>% 
    mutate(asv_core = case_when(host_p>th ~ "Core",
                                .default = "Non-core"))
  core_names <- unique(data_asv_p$asv_core)
  
  
  # loop for two groups
  for (v in core_names) {
    
    data_asv_group <- data_asv_p %>% 
      filter(asv_core == v)
    
    ##### observed network
    # finding modules
    modules_observed <- fun_modularity_analysis(data_asv_group) %>% 
      mutate(threshold=th)
    modules_table_two_groups <- rbind(modules_table_two_groups, modules_observed)
    
    
    # calculating similarity in modules
    modules_similarity <- fun_modules_similarity(modules_observed)
    # saving results in one table
    modules_similarity_two_groups <- append(modules_similarity_two_groups, modules_similarity)
    
    
    ##### assembly analysis
    
    # ASV pool - ASVs of core/non-core group in the specific threshold
    asv_pool <- data_asv_group %>% 
      distinct(asv_ID)
    
    data_assembly <- data_asv_group %>% 
      mutate(reads=1) %>% 
      distinct(host_ID, asv_ID, reads) %>% 
      spread(asv_ID, reads, fill = 0) %>% 
      column_to_rownames("host_ID") %>% 
      as.matrix()
    
    # filtering phylogenetic distance of only the ASV pool
    asv_distance2 <- asv_distance[rownames(asv_distance) %in% asv_pool$asv_ID , colnames(asv_distance) %in% asv_pool$asv_ID]
    
    # calculating raup-cric
    raupc <- as.matrix(vegan::raupcrick(data_assembly, nsimul = n_sim))
    # calculating betaNTI
    assembly <- iCAMP::qpen(data_assembly, pd = asv_distance2, RC=raupc, rand=n_sim, sig.bNTI = 2, sig.rc = 0.95, ab.weight=F, nworker = 4)
    
    assembly_result <- assembly$result %>% 
      mutate(sample1 = as.double(sample1), sample2 = as.double(sample2)) %>% 
      left_join(modules_observed %>% 
                  distinct(host_ID, host_group, grid), by=c("sample1"="host_ID")) %>% 
      dplyr::rename(host_group1=host_group, grid1=grid) %>% 
      left_join(modules_observed %>% 
                  distinct(host_ID, host_group, grid), by=c("sample2"="host_ID")) %>% 
      dplyr::rename(host_group2=host_group, grid2=grid) %>% 
      mutate(threshold = th, asv_core = v)
    
    # saving the assembly results in the final table
    assembly_two_groups <- rbind(assembly_two_groups, assembly_result)
    
  }  
}
```

# Network exploration

```{r, network_exploration}
# number of ASVs
asv_total <- length(unique(modules_table_two_groups$asv_ID))
host_total <- modules_table_two_groups %>% group_by(threshold, asv_core) %>% summarise(total_host = n_distinct(host_ID))

group_n <- modules_table_two_groups %>% 
  group_by(threshold, asv_core) %>% 
  summarise(asv_n = n_distinct(asv_ID), host_n = n_distinct(host_ID)) %>% 
  mutate(asv_prop = asv_n/asv_total)

# connectance
interaction_n <- modules_table_two_groups %>% 
  count(threshold, asv_core)
group_n %<>% left_join(interaction_n, by=c("threshold","asv_core")) %>% 
  mutate(connectance = n / (asv_n*host_n))

# number of modules
modules_n <- modules_table_two_groups %>% 
  group_by(threshold, asv_core) %>% 
  summarise(modules_n = n_distinct(host_group))

# mean modules size
modules_size <- modules_table_two_groups %>% 
  distinct(threshold, asv_core, host_group, host_ID) %>% 
  count(threshold, asv_core, host_group) %>% 
  group_by(threshold, asv_core) %>% 
  summarise(modules_size = mean(n), largest_module = max(n)) %>% 
  left_join(host_total) %>% 
  mutate(largest_module_prop = largest_module/total_host)

modules_n %<>% left_join(modules_size, by=c("threshold","asv_core")) 

# final table of network properties
final_network_properties <- group_n %>% 
  left_join(modules_n, by=c("threshold","asv_core"))

# plotting
g <- final_network_properties %>% 
  select(-n, -host_n, -asv_n, -total_host, -largest_module) %>% 
  pivot_longer(cols=asv_prop:largest_module_prop, names_to="measure", values_to="value") %>% 
  mutate(measure = case_when(measure=="asv_prop" ~ "Prop. of ASVs",
                             measure=="connectance" ~ "Connectance",
                             measure=="modules_n" ~ "No. of Modules",
                             measure=="modules_size" ~ "Avg. Module Size",
                             measure=="largest_module_prop" ~ "Largest Module")) %>% 
  mutate(measure = factor(measure, levels = c("Prop. of ASVs","Connectance","No. of Modules","Avg. Module Size","Largest Module"))) %>%
  ggplot(aes(x=threshold, y=value, color=asv_core)) +
  geom_line() +
  geom_vline(xintercept = 0.2, linetype = "dotted", color = "grey")+
  facet_wrap(~measure, scales="free", nrow = 1) +
  theme_bw() +
  theme(axis.text = element_text(size = 9, color = 'black', angle = 90, vjust = 0.5, hjust=1), 
        title = element_text(size = 14), 
        legend.position = "none",
        strip.text = element_text(size=10, color = 'black'), 
        strip.background = element_rect(color = "grey80", size = 1), 
        panel.grid = element_blank(), 
        panel.background = element_rect(colour = "black"), 
        aspect.ratio = 1.7
        ) +
  scale_color_manual(values=group.colors) +
  labs(x="Threshold", y="Value", color="Microbial Group")
print(g)

```



# assembly processes analysis

```{r, assembly_processes, eval=FALSE, include=FALSE}
# whether two hosts from the same module?
assembly_final <- assembly_two_groups %>% 
  mutate(same_module = ifelse(host_group1==host_group2, "Same Module","Different Module"))

# renaming the processes
assembly_final %<>% mutate(process = case_when(process=="Heterogeneous.Selection" ~ "Heterogeneous Selection",
                                               process=="Homogeneous.Selection" ~ "Homogeneous Selection",
                                               process=="Dispersal.Limitation" ~ "Dispersal Limitation",
                                               process=="Homogenizing.Dispersal" ~ "Homogenizing Dispersal",
                                               .default = "Ecological Drift"))

# summary
assembly_summary <- assembly_final %>% 
  count(threshold, asv_core, same_module, process)

assembly_summary_total_module <- assembly_summary %>% 
  group_by(threshold, asv_core, same_module) %>% 
  summarise(n_total = sum(n))


#plotting process ratio between modules
g <- assembly_summary %>% 
  left_join(assembly_summary_total_module, by=c("threshold","asv_core","same_module")) %>% 
  mutate(n_p = n/n_total) %>% 
  ggplot(aes(fill=process, x=threshold, y=n_p)) +
  geom_area() +
  facet_grid(cols = vars(asv_core), rows = vars(same_module)) +
  geom_vline(xintercept = 0.2, linetype = "dotted", color = "grey")+
  scale_y_continuous(labels = scales::percent) +
  theme_bw() +
  theme(axis.text = element_text(size = 9, color = 'black', angle = 90, vjust = 0.5, hjust=1), 
        title = element_text(size = 14),
        strip.text = element_text(size=12, color = 'black'), 
        strip.background = element_rect(color = "grey80", size = 1), 
        panel.grid = element_blank(), 
        panel.background = element_rect(colour = "black")
  ) +
  scale_fill_manual(values=c("#e0ac69","#f2e8cf","#bcbddc","#807dba")) +
  labs(x="Threshold", y="Percentage", fill="Process")
print(g)

```



# Modules similarity across land use change gradient

## Correlations between variables

```{r, variables_correlation}
# reading grid attributes
grids_similarity_attr <- read_csv("../data/data_processed/village_summary.csv")
rownames(grids_similarity_attr) <- grids_similarity_attr$grid_village

# correlations between variables
library(psych)
print(psych::pairs.panels(grids_similarity_attr %>%  select(-grid_village,-village,-grid), ellipses = F, lm = F))
cat('\n','\n')
```

## RDA analysis

```{r}
# RDA for modules
# the analysis is done separately for each core/non-core group at specific threshold

par_names <- expand_grid(core_th, core_names)
anova_model_all <- NULL
anova_rda_all <- NULL
set.seed(123)

# loop for all combinations: core/non-core X thresholds
for (i in 1:nrow(par_names)) {
  
  modules_similarity <- modules_similarity_two_groups[[i]]
  # filtering the largest module
  largest_module <- which.max(colSums(modules_similarity))
  modules_similarity2 <- modules_similarity[,-largest_module]
  
  grid_names <- rownames(modules_similarity2)
  
  # filtering matrix to the existing grids
  grids_similarity_attr2 <- grids_similarity_attr %>%  
    filter(grid_village %in% grid_names & !grepl("village", grid_village))
  modules_similarity2 <- modules_similarity2[rownames(grids_similarity_attr2),]
  
  # number of hosts in each site
  n=data_asv %>% group_by(village,grid) %>% summarise(n=n_distinct(host_ID))
  grids_similarity_attr2 %<>% left_join(n, by=c("village","grid"))
  
  # standardizing variables to zero mean and unit variance
  grids_similarity_attr2[4:10] <- decostand(grids_similarity_attr2[4:10], 'standardize')
  
  
  ##### tb-RDA analysis
  
  # hellinger transformation
  modules_similarity_hell <- decostand(modules_similarity2, 'hell')
  
  # RDA analysis
  rda_result <- rda(modules_similarity_hell ~ veg_PC1+veg_PC2+dist_to_village+elevation+Condition(n), data=grids_similarity_attr2, na.action = na.omit)
  # **Condition(n) - control for the number of host in each village-grid.
  
  # variation explained
  constrained_eig <- t(as.data.frame(rda_result$CCA$eig/rda_result$tot.chi*100))  # for RDAs
  # R2
  R2.obs <- RsquareAdj (rda_result)$adj.r.squared
  
  # significance test for the whole model
  anova_model <- anova(rda_result)[1,] %>% 
    mutate(R2.adj = R2.obs, `Pr(>F)` = as.double(`Pr(>F)`), asv_core = par_names[[i,2]], core_th = par_names[[i,1]]) %>% 
    mutate(sig = ifelse(`Pr(>F)` <= 0.05, 1, 0))
  
  anova_model_all <- rbind(anova_model_all, anova_model)
  
  # significance test for all variables
  anova_rda <- anova(rda_result, by = "margin", permutations = 999) %>% 
    mutate(asv_core = par_names[[i,2]], core_th = par_names[[i,1]]) 
  # correct for multiple tests
  anova_rda$`Pr(>F)` <- p.adjust (anova_rda$`Pr(>F)`, method = 'holm') 
  anova_rda %<>% mutate(`Pr(>F)` = as.double(`Pr(>F)`), sig = ifelse(`Pr(>F)` <= 0.05, 1, 0)) 
  
  anova_rda_all <- rbind(anova_rda_all, anova_rda)
  
}

######
# plotting

# arranging the tables
anova_model_all_plot <- anova_model_all %>% 
  mutate(R2.adj = ifelse(R2.adj<0,0,R2.adj))

anova_rda_all_plot <- anova_rda_all %>% 
  rownames_to_column("var") %>% 
  left_join(anova_model_all_plot %>% filter(sig==1) %>% select(core_th,asv_core,R2.adj), by=c("asv_core","core_th")) %>% 
  filter(sig == 1 & !is.na(R2.adj)) %>% 
  group_by(core_th, asv_core, R2.adj) %>%
  summarise(var_combined = paste(var, collapse = "\n"), .groups = "drop")

g <- anova_model_all_plot %>% 
  ggplot(aes(y=R2.adj, x=core_th, color = asv_core)) +
  geom_vline(xintercept = 0.2, linetype = "dotted", color = "grey")+
  geom_point(aes(shape = as.factor(sig), size = 3)) +
  geom_line()+
  scale_size(guide="none") +
  scale_shape_manual(values = c(1, 16), guide="none") +
  scale_y_continuous(limits = c(0, 0.2)) +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), 
        title = element_text(size = 14), 
        #legend.position = "bottom",
        panel.grid = element_blank(), 
        panel.border = element_rect(color = "black"), 
        aspect.ratio = 0.5
  ) +
  scale_color_manual(values = group.colors) +
  labs(x = "Threshold", y = bquote("tb-RDA " * R[Adjusted]^2 * ""), color = "Microbial Group")
print(g)
cat('\n','\n')


# printing the tables
print(knitr::kable(anova_model_all))
cat('\n','\n')

print(knitr::kable(anova_rda_all))
cat('\n','\n')

```

# Relative abundance microbial families across modules

```{r, microbiome taxonomy}
# microbial families relative abundance across modules

# microbes taxonomy
tax <- read_delim("../data/data_raw/data_microbiome/ASVs_taxonomy_new.tsv") %>% 
  dplyr::rename(asv_ID = ASV)

data_asv_tax <- modules_table_two_groups %>% 
  left_join(tax, by="asv_ID")

data_tax_final = NULL
core_th2 <- seq(0.1, 0.5, by=0.1)

# loop for different thresholds
for (t in core_th2) {
  
  data_asv_tax_p <- data_asv_tax %>% 
    filter(near(threshold , t))
  
  # total number of reads of each core/non-core group
  total_reads_groups <- data_asv_tax_p %>% 
    distinct(host_ID, asv_core, total_reads) %>% 
    group_by(asv_core) %>% 
    summarise(n_total = sum(total_reads))
  
  # relative abundance of each family
  most_abu_family <- data_asv_tax_p %>%
    mutate(reads_a = reads*total_reads) %>%
    group_by(asv_core, Family) %>%
    summarise(n= sum(reads_a)) %>%
    left_join(total_reads_groups, by="asv_core") %>%
    mutate(n_p = n/n_total) 
  
  # selecting the 8 most abundant families
  most_abu_family_8 <- most_abu_family %>% 
    filter(!is.na(Family)) %>% 
    ungroup() %>% 
    slice_max(by = asv_core, order_by = n_p, n = 8) %>% 
    mutate(p = "top") %>% 
    select(asv_core, Family, p)
  
  # selecting the 5 largest modules
  largest_modules <- modules_table_two_groups %>%
    filter(near(threshold , t)) %>%
    group_by(asv_core, host_group) %>% 
    summarise(n = n_distinct(host_ID)) %>% 
    ungroup() %>%
    slice_max(by = asv_core, order_by = n, n = 5, with_ties = FALSE) %>% 
    mutate(host_group2 = rep(1:5,2), p = "top")
  
  # combining modules table with the list of largest modules
  modules_top <- modules_table_two_groups %>% 
    filter(near(threshold , t)) %>% 
    left_join(largest_modules, by=c("asv_core","host_group")) %>% 
    mutate(module = factor(ifelse(is.na(p),"Other",host_group2))) %>%
    distinct(host_ID,host_group,host_group2,asv_core, module) 
  
  # total reads in the 5 largest modules
  total_reads_modules <-  modules_table_two_groups %>% 
    filter(near(threshold , t)) %>% 
    distinct(host_ID, asv_core, host_group, total_reads)  %>% 
    left_join(modules_top %>% distinct(asv_core,host_group,module), by=c("asv_core","host_group")) %>% 
    group_by(asv_core, module) %>% summarise(n_total = sum(total_reads))
  
  # relative abundance of the 8 most abundant families in the 5 largest modules
  most_abu_family_p <- data_asv_tax_p %>% 
    left_join(modules_top, by=c("asv_core","host_ID")) %>% 
    mutate(reads_a = reads*total_reads) %>%
    group_by(asv_core, Family, module) %>%
    summarise(n= sum(reads_a)) %>%
    left_join(total_reads_modules, by=c("asv_core","module")) %>%
    mutate(n_p = n/n_total) %>% 
    left_join(most_abu_family_8, by=c("asv_core","Family")) %>% 
    mutate(p = case_when(p=="top" ~ Family,
                         is.na(Family) ~ ".NA",
                         is.na(p) ~ ".Other")) %>% 
    mutate(threshold = t)
  
  data_tax_final <- rbind(data_tax_final, most_abu_family_p)
}

# plotting
colors <- c("grey40","grey80","darkblue","#ffd60a",brewer.pal(n = 12, name = "Paired"))

g <- data_tax_final %>% 
  group_by(threshold, asv_core, module, p) %>% 
  summarise(n_p = sum(n_p)) %>% 
  ggplot(aes(fill=p, x=module, y=n_p)) +
  geom_bar(position="fill", stat="identity", width=0.8) +
  facet_grid(cols = vars(threshold), rows = vars(asv_core), scales = "free_x") +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), 
        title = element_text(size = 14),
        strip.text = element_text(size=12, color = 'black'), 
        axis.text.x = element_text(angle = 60, vjust = 1.1, hjust=1),
        panel.grid = element_blank(), 
        panel.background = element_rect(colour = "black"), 
        legend.key.size = unit(0.5, "cm")
  ) +
  scale_fill_manual(values=colors) +
  labs(x="Modules", y="Relative Abundance", fill = "Family")

print(g)
cat('\n','\n')

```

# microbiome taxonomy across thresholds

```{r, taxonomy}
# microbiome taxonomy across core thresholds

tax <- read_delim("../data/data_raw/data_microbiome/ASVs_taxonomy_new.tsv") %>% 
  dplyr::rename(asv_ID = ASV)

# calculating total reads for each ASV for core/non-core across all thresholds
data_asv_unique <- modules_table_two_groups %>% 
  mutate(reads = reads*total_reads) %>% 
  group_by(asv_ID, asv_core, threshold) %>% 
  summarise(total_reads = sum(reads)) %>% 
  left_join(tax, by='asv_ID')

# aggregating to the family level
# filtering ASVs with no identified family
data_asv_family <- data_asv_unique %>% 
  filter(!is.na(Family)) %>% 
  group_by(Family, asv_core, threshold) %>% 
  summarise(total_reads = sum(total_reads))

# total reads for all core/non-core
total_reads_group <- data_asv_family %>% 
  group_by(asv_core, threshold) %>% 
  summarise(total_reads_core = sum(total_reads))

# calculating family relative reads for every group and threshold 
data_asv_family %<>% left_join(total_reads_group, by=c("asv_core","threshold")) %>% 
  mutate(relative_reads = total_reads/total_reads_core)

# bray-curtis at the family level for all thresholds
results_bray <- NULL
for (t2 in core_th) {
  
  dat <- data_asv_family %>% 
    filter(near(threshold , t2)) 
  
  dat_matrix <- dat %>% 
    pivot_wider(names_from = Family, values_from = relative_reads, id_cols = asv_core) %>% 
    column_to_rownames("asv_core") %>% 
    as.matrix()
  dat_matrix[is.na(dat_matrix)] <- 0
  
  # calculating dis-similarity!
  results <- data_frame(threshold = t2,
                        bray_curtis = vegan::vegdist(dat_matrix, method = "bray"))
  results_bray <- rbind(results_bray, results)
}

# plotting bray-curtis dissimilarity
g1 <- results_bray %>% 
  ggplot(aes(x=threshold, y=bray_curtis)) +
  geom_line() +
  geom_vline(xintercept = 0.2, linetype = "dotted", color = "grey")+
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), 
        title = element_text(size = 14), 
        legend.position = "none",
        strip.text = element_text(size=10, color = 'black'), 
        strip.background = element_rect(color = "grey80", size = 1), 
        panel.grid = element_blank(), 
        panel.background = element_rect(colour = "black"), 
        aspect.ratio = 1
  ) +
  labs(x="Threshold", y="Bray-Curtis Dissimilarity")
print(g1)
cat('\n','\n')


# plotting family relative abundance for core/non-core at all thresholds

colors <- c("black","grey40","grey80","#15616d","#fefae0","darkblue","#ffd60a",brewer.pal(n = 12, name = "Paired"))

# plotting only families with relative abundance > 1%
g2 <- data_asv_family %>% 
  filter(relative_reads >= 0.01) %>% 
  mutate(asv_core = case_when(asv_core=="Core" ~ "C",
                              asv_core=="Non-core" ~ "N")) %>% 
  ggplot(aes(fill=Family, x=asv_core, y=relative_reads)) +
  geom_bar(position="fill", stat="identity") +
  facet_wrap(~threshold, scales = "free_x", ncol = 5) +
  theme_bw() +
  theme(axis.text = element_text(size = 10, color = 'black'), 
        title = element_text(size = 14),
        strip.text = element_text(size=12, color = 'black'), 
        panel.grid = element_blank(), 
        panel.background = element_rect(colour = "black"), 
        legend.key.size = unit(0.5, "cm")
  ) +
  scale_fill_manual(values=colors) +
  labs(x="Microbial Group", y="Relative Abundance", fill = "Family")
print(g2)
cat('\n','\n')

```

