---
title: "Differential assembly of core and non-core host-microbe network structures along a land-use change gradient"
subtitle: "Analysis and results - a single core/non-core threshold"
author: "Matan Markfeld"
date: "Last edit: 2025-04-03"
output: 
  html_document:
    toc: true
    toc_float: true
    collapse: false
    number_sections: true
    code_folding: hide
    highlight: tango
---
```{r setup, include=FALSE}
knitr::opts_chunk$set(results = "asis", message=FALSE, warning=FALSE, cache=TRUE, eval = TRUE, dev = c('png','pdf'), out.width = '100%', out.height='60%')
```


```{r load libraries, echo=FALSE, message=FALSE, warning=FALSE}
library(tidyverse)
library(magrittr)
library(ggplot2)
library(vegan)
library(igraph)
library(reshape2)
library(infomapecology)
library(ape)
library(patchwork)
library(iCAMP)
library(ggrepel)

rm(list=ls())
```

```{r}
source("functions.R", local = knitr::knit_global())
```

```{r, microbiome data}
# reading the microbiome data
# working only with Rattus from the three villages

data_asv <- read_csv("../data/data_processed/data_asv_rra0.001_p0.01_th5000_all.csv")

# reading the phylogenetic tree
best_tree <- readRDS(file = "../data/data_processed/phylo_tree_rra0.001_p0.01.rds")
phylo_tree <- best_tree$tree 
# ASVs phylogenetic distance
asv_distance <- ape::cophenetic.phylo(phylo_tree)
```

```{r, main analysis, include=FALSE, results='hide'}
# first of all, run this chunk of code.
# this code does all the analysis using the functions in "functions".
# the output of the code is variables holding the final results tables and figures (the output of the functions used)

# ------------------------------------------------------------------------------
# thresholds for core
core_th <- 0.2
# number of simulations to the assembly analysis
n_sim <- 2

# group colors
group.colors <- c(Core = "#d95f02", 'Non-core' = "#034e7b")

# final tables 
modules_similarity_two_groups <- NULL
modules_table_two_groups <- NULL
asv_degree_distribution_two_groups <- NULL
modules_two_groups <- NULL
modules_size_two_groups <- NULL
modules_grid_two_groups <- NULL
assembly_two_groups <- NULL
# ------------------------------------------------------------------------------

# defining core/non-core ASVs based on the threshold (th) 
data_asv_p <- data_asv %>% 
  mutate(asv_core = case_when(host_p>core_th ~ "Core",
                              .default = "Non-core"))

core_names <- unique(data_asv_p$asv_core)

# loop for two groups
for (v in core_names) {
  
  data_asv_group <- data_asv_p %>% 
    filter(asv_core == v)
  
  # calculating ASVs degree
  asv_degree <- data_asv_group %>%
    group_by(asv_ID) %>%
    summarise(n = n_distinct(host_ID)) %>%
    dplyr::rename(asv_degree = n)
  data_asv_group %<>% left_join(asv_degree, by="asv_ID")
  
  
  ##### observed network
  # finding modules
  modules_observed <- fun_modularity_analysis(data_asv_group) %>% 
    mutate(threshold=core_th)
  modules_table_two_groups <- rbind(modules_table_two_groups, modules_observed)
  
  # calculating similarity in modules
  modules_similarity <- fun_modules_similarity(modules_observed)
  # saving results in one table
  modules_similarity_two_groups <- append(modules_similarity_two_groups, modules_similarity)
  
  # ASVs degree distribution
  asv_degree_distribution <- fun_asv_degree_distribution(data_asv_group)
  asv_degree_distribution_two_groups <- append(asv_degree_distribution_two_groups, asv_degree_distribution)
  
  # figure of modules 
  modules <- fun_modules(modules_observed, v)
  modules_two_groups <- append(modules_two_groups, modules)
  
  # modules size
  modules_size <- fun_module_size(modules_observed)
  modules_size_two_groups <- append(modules_size_two_groups, modules_size)
  
  # Number of land uses per module
  modules_grid <- fun_module_grid(modules_observed)
  modules_grid_two_groups <- append(modules_grid_two_groups, modules_grid)
  
  
  ##### assembly analysis
  # ASV pool - ASVs of core/non-core group in the specific threshold
  asv_pool <- data_asv_group %>% 
    distinct(asv_ID)
  
  data_assembly <- data_asv_group %>% 
    mutate(reads=1) %>% 
    distinct(host_ID, asv_ID, reads) %>% 
    spread(asv_ID, reads, fill = 0) %>% 
    column_to_rownames("host_ID") %>% 
    as.matrix()
  
  # filtering phylogenetic distance of only the ASV pool
  asv_distance2 <- asv_distance[rownames(asv_distance) %in% asv_pool$asv_ID , colnames(asv_distance) %in% asv_pool$asv_ID]
  
  # calculating raup-cric
  raupc <- as.matrix(vegan::raupcrick(data_assembly, nsimul = n_sim))
  # calculating betaNTI
  assembly <- iCAMP::qpen(data_assembly, pd = asv_distance2, RC=raupc, rand=n_sim, sig.bNTI = 2, sig.rc = 0.95, ab.weight=F, nworker = 4)
  
  assembly_result <- assembly$result %>% 
    mutate(sample1 = as.double(sample1), sample2 = as.double(sample2)) %>% 
    left_join(modules_observed %>% 
                distinct(host_ID, host_group, grid), by=c("sample1"="host_ID")) %>% 
    dplyr::rename(host_group1=host_group, grid1=grid) %>% 
    left_join(modules_observed %>% 
                distinct(host_ID, host_group, grid), by=c("sample2"="host_ID")) %>% 
    dplyr::rename(host_group2=host_group, grid2=grid) %>% 
    mutate(threshold = core_th, asv_core = v)
  
  # saving the assembly results in the final table
  assembly_two_groups <- rbind(assembly_two_groups, assembly_result)
  
}
```


# Network exploration {.tabset}

```{r, network_exploration}

# loop for two groups
for (i in 1:2) {
  
  cat('##',core_names[i],'{.tabset}','\n','\n')
  
  cat('### ASVs degree distribution','\n')
  
  print(asv_degree_distribution_two_groups[[i]])
  cat('\n','\n')
  
  # calculating network properties
  connectance_data <- modules_table_two_groups %>% 
    filter(asv_core == core_names[i])
  
  cat('No. of hosts: ', length(unique(connectance_data$host_ID)) ,'\n','\n')
  cat('No. of ASVs: ', length(unique(connectance_data$asv_ID)) ,'\n','\n')
  
  asv_mean_degree <- connectance_data %>% distinct(asv_ID,asv_degree) %>% summarise(mean = mean(asv_degree), sd = sd(asv_degree))
  cat('ASV mean degree: ', asv_mean_degree$mean, 'ASV sd degree: ', asv_mean_degree$sd,'\n','\n')
  
  host_mean_degree <- connectance_data %>% count(host_ID) %>% summarise(mean = mean(n), sd = sd(n))
  cat('Host mean degree: ', host_mean_degree$mean, 'Host sd degree: ', host_mean_degree$sd,'\n','\n')
  
  cat('Connectance: ', nrow(connectance_data) / (length(unique(connectance_data$host_ID)) * length(unique(connectance_data$asv_ID))) ,'\n','\n')
  
  no_modules <- modules_table_two_groups %>% 
    filter(asv_core == core_names[i]) %>%
    summarise(n = n_distinct(host_group))
  cat('Number of modules: ', no_modules$n,'\n','\n')
  
  modules_sizes <- modules_table_two_groups %>% 
    filter(asv_core == core_names[i]) %>% 
    group_by(asv_core, host_group) %>% 
    summarise(n = n_distinct(host_ID)) %>% 
    summarise(mean = mean(n), sd = sd(n))
  cat('Module mean size: ', modules_sizes$mean, 'Module sd size: ', modules_sizes$sd,'\n','\n')
  
  
  
  cat('### Modules','\n')
  cat('The color indicates number of host individuals in the module / total number of hosts in the whole grid [%]','\n','\n')
  
  print(modules_two_groups[[i]])
  cat('\n','\n')
  
  cat('### Modules size','\n')
  print(modules_size_two_groups[[i]])
  cat('\n','\n')
  
  cat('### No. of land uses','\n')
  print(modules_grid_two_groups[[i]])
  cat('\n','\n')
}

# final modules figure - core + non-core together
cat('## Modules','\n')
modules_two_groups[[1]]+modules_two_groups[[2]] + plot_layout(nrow=1,ncol=2,guides='collect',axes = "collect_y",axis_titles = "collect") &
  theme(legend.position='bottom')
cat('\n','\n')
```


# assembly processes analysis

```{r, assembly_processes}
# combining final results
assembly_final <- assembly_two_groups %>% 
  mutate(same_module = ifelse(host_group1==host_group2, "Same","Different"))


# renaming the processes
assembly_final %<>% mutate(process = case_when(process=="Heterogeneous.Selection" ~ "Heterogeneous Selection",
                                               process=="Homogeneous.Selection" ~ "Homogeneous Selection",
                                               process=="Dispersal.Limitation" ~ "Dispersal Limitation",
                                               process=="Homogenizing.Dispersal" ~ "Homogenizing Dispersal",
                                               .default = "Drift"))

# summary
assembly_summary <- assembly_final %>% 
  count(asv_core, same_module, process)

assembly_summary_total <- assembly_summary %>% 
  group_by(asv_core) %>% 
  summarise(n_total = sum(n))

assembly_summary_total_module <- assembly_summary %>% 
  group_by(asv_core, same_module) %>% 
  summarise(n_total = sum(n))

# plotting process ratio between groups
g1 <- assembly_summary %>% 
  group_by(asv_core, process) %>% 
  summarise(n = sum(n)) %>% 
  left_join(assembly_summary_total, by=c("asv_core")) %>% 
  mutate(n_p = n/n_total) %>% 
  ggplot(aes(fill=process, x=asv_core, y=n_p)) +
  geom_bar(position="fill", stat="identity") +
  scale_y_continuous(labels = scales::percent) +
  geom_text(aes(label = paste0(round(n_p*100,1),"%")), 
            position = position_stack(vjust = 0.5), size = 3) +
  theme_bw() +
  theme(axis.text = element_text(size = 11, color = 'black', angle = 90, vjust = 0.5, hjust=1), 
        title = element_text(size = 14),
        strip.text = element_text(size=12, color = 'black'),
        strip.background = element_rect(color = "grey80", size = 1), 
        panel.grid = element_blank(), 
        panel.background = element_rect(colour = "black")
  ) +
  scale_fill_manual(values=c("#e0ac69","#f2e8cf","#bcbddc","#807dba")) +
  labs(x="ASVs Groups", y="Percentage")
print(g1)

# plotting process ratio between modules
g2 <- assembly_summary %>% 
  left_join(assembly_summary_total_module, by=c("asv_core","same_module")) %>% 
  mutate(n_p = n/n_total) %>% 
  ggplot(aes(fill=process, x=same_module, y=n_p)) +
  geom_bar(position="fill", stat="identity") +
  facet_wrap(~asv_core) +
  scale_y_continuous(labels = scales::percent) +
  geom_text(aes(label = paste0(round(n_p*100,1),"%")), 
            position = position_stack(vjust = 0.5), size = 3) +
  theme_bw() +
  theme(axis.text = element_text(size = 9, color = 'black', angle = 90, vjust = 0.5, hjust=1), 
        title = element_text(size = 14),
        strip.text = element_text(size=12, color = 'black'), 
        strip.background = element_rect(color = "grey80", size = 1), 
        panel.grid = element_blank(), 
        panel.background = element_rect(colour = "black")
  ) +
  scale_fill_manual(values=c("#e0ac69","#f2e8cf","#bcbddc","#807dba")) +
  labs(x="Hosts from the Same/Different Modules", y="Percentage", fill="Process")
print(g2)
```


# Modules similarity across land use change gradient

## Correlations between variables

```{r, variables_correlation}
# reading grid attributes
grids_similarity_attr <- read_csv("../data/data_processed/village_summary.csv")
rownames(grids_similarity_attr) <- grids_similarity_attr$grid_village

# correlations between variables
library(psych)
print(psych::pairs.panels(grids_similarity_attr %>%  select(-grid_village,-village,-grid), ellipses = F, lm = F))
cat('\n','\n')
```

```{r}
# RDA for modules

anova_model_all <- NULL
anova_rda_all <- NULL
rda_figs <- list()
set.seed(123)

# loop for two asv groups
for (i in 1:2) {
  
  modules_similarity <- modules_similarity_two_groups[[i]]
  # filtering the largest module
  largest_module <- which.max(colSums(modules_similarity))
  modules_similarity2 <- modules_similarity[,-largest_module]
  
  grid_names <- rownames(modules_similarity2)
  
  # filtering matrix to the existing grids
  grids_similarity_attr2 <- grids_similarity_attr %>%  
    filter(grid_village %in% grid_names & !grepl("village", grid_village))
  modules_similarity2 <- modules_similarity2[rownames(grids_similarity_attr2),]
  
  # number of hosts in each site
  n=data_asv %>% group_by(village,grid) %>% summarise(n=n_distinct(host_ID))
  grids_similarity_attr2 %<>% left_join(n, by=c("village","grid"))
  
  # standardizing variables to zero mean and unit variance
  grids_similarity_attr2[4:10] <- decostand(grids_similarity_attr2[4:10], 'standardize')
  
  
  ##### tb-RDA analysis
  
  # hellinger transformation
  modules_similarity_hell <- decostand(modules_similarity2, 'hell')
  
  # RDA analysis
  rda_result <- rda(modules_similarity_hell ~ veg_PC1+veg_PC2+dist_to_village+elevation+Condition(n), data=grids_similarity_attr2, na.action = na.omit)
  # **Condition(n) - control for the number of host in each village-grid.
  
  # variation explained
  constrained_eig <- t(as.data.frame(rda_result$CCA$eig/rda_result$tot.chi*100))  # for RDAs
  # R2
  R2.obs <- RsquareAdj (rda_result)$adj.r.squared
  
  # significance test for the whole model
  anova_model <- anova(rda_result)[1,] %>% 
    mutate(R2.adj = R2.obs, asv_core = core_names[i])
  
  anova_model_all <- rbind(anova_model_all, anova_model)
  
  # significance test for all variables
  anova_rda <- anova(rda_result, by = "margin", permutations = 999) %>% 
    mutate(asv_core = core_names[i])
  anova_rda$`Pr(>F)` <- p.adjust (anova_rda$`Pr(>F)`, method = 'holm')  
  
  anova_rda_all <- rbind(anova_rda_all, anova_rda)
  
  
  ##### plotting
  # samples
  rda_samples <- as.data.frame(scores(rda_result)$sites) %>% 
    rownames_to_column("grid_village") %>% 
    left_join(grids_similarity_attr2[c("grid_village","village","grid")], by="grid_village") %>% 
    mutate(asv_core = core_names[i]) %>% 
    mutate(grid = factor(grid, levels = c("semi-intact_forest","secondary_forest","brushy_regrowth","agriculture","agroforest","flooded_rice","village"))) 
  
  # variables
  rda_var <- as.data.frame(rda_result$CCA$biplot)
  rownames(rda_var) <- c("Vegetation1","Vegetation2","Distance","Elevation")
  
  # plot
  lu.color <- c("#005824","#238b45","#41ae76","#66c2a4","#99d8c9","#ccece6")
  g <- rda_samples %>% 
    ggplot(aes(RDA1, RDA2, color=grid)) +
    geom_hline(yintercept = 0, linetype = "dotted", color = "grey") +
    geom_vline(xintercept = 0, linetype = "dotted", color = "grey") +
    geom_point(size = 4) +
    geom_segment(
      data = rda_var, 
      inherit.aes = FALSE, 
      alpha = 0.8,
      aes(x = 0, y = 0, xend = (RDA1 * 0.85), yend = (RDA2 * 0.85)), 
      arrow = arrow(length = unit(0.15, "cm"))
    ) +
    geom_text_repel(
      data = rda_var, 
      inherit.aes = FALSE, 
      aes(x = RDA1, y = RDA2, label = rownames(rda_var)), 
      size = 4, 
      box.padding = 0.5,
      point.padding = 0.5,
      segment.colour = NA
    ) +
    theme_bw() +
    coord_cartesian(
      xlim = range(rda_samples$RDA1) * c(1.3, 1.2), 
      ylim = range(rda_samples$RDA2) * c(1.3, 1.2), 
      clip = 'off'
    ) +
    theme(
      panel.border = element_rect(colour = "black", size = 1), 
      panel.grid = element_blank(), aspect.ratio = 1
    ) +
    scale_color_manual(values=lu.color) +
    labs(
      x = paste("RDA1 (", round(constrained_eig[1], 2), "%)", sep = ""), 
      y = paste("RDA2 (", round(constrained_eig[2], 2), "%)", sep = ""), 
      title = paste(core_names[i]), 
      color = "Land Use", 
      shape = "Village"
    )
  
  rda_figs <- append(rda_figs, list(g))
}

# printing tables
print(knitr::kable(anova_model_all))
cat('\n','\n')

print(knitr::kable(anova_rda_all))
cat('\n','\n')

# printing final figure
rda_figs[[1]] + rda_figs[[2]] + plot_layout(ncol=2, nrow=1, guides='collect') & theme(legend.position='right')
cat('\n','\n')
```
